
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Advanced usage &#8212; SCOOTI 0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="SCOOTI Settings" href="scootisettings.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="advanced-usage">
<h1>Advanced usage<a class="headerlink" href="#advanced-usage" title="Permalink to this heading">¶</a></h1>
<section id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this heading">¶</a></h2>
<p>In the previous section, we have demonstrated how users can quickly set up and start using <strong>SCOOTI</strong>. In this section, we will show you how to take advantage of the flexibility of <strong>SCOOTI</strong> to support your research.</p>
</section>
<section id="identifying-up-and-down-regulated-genes">
<h2>Identifying up- and down-regulated genes<a class="headerlink" href="#identifying-up-and-down-regulated-genes" title="Permalink to this heading">¶</a></h2>
<p>SCOOTI relies on the results of flux prediction generated by metabolic modeling methods to infer objectives. Therefore, to prepare context-specific flux prediction based on omics data, we need to identify up- and down-regulated genes from data. The method is mainly designed for integrating transcriptomics and proteomics. Here, we provide three different ways users can consider for their own models.</p>
<p>No matter what types of method users are interested in, the initiation of the Python class <cite>findSigGenes</cite> is required to apply the analysis functions.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># instance of the python class</span>
<span class="n">fr</span> <span class="o">=</span> <span class="n">findRegulators</span><span class="p">(</span><span class="s1">&#39;./data/scEmbryo/GSE136714/single_cell/&#39;</span><span class="p">)</span>
</pre></div>
</div>
<section id="convert-single-cell-data-from-table-to-10x-format">
<h3>Convert single-cell data from table to 10x format<a class="headerlink" href="#convert-single-cell-data-from-table-to-10x-format" title="Permalink to this heading">¶</a></h3>
<p>Single-cell data provide the information of expression uncertainty. No matter what type of single-cell data, such as scRNA-seq, single nucleus RNA-seq, and ATAC-seq, <cite>findSigGenes</cite> leveraging <cite>scanpy</cite> and <cite>anndata</cite> can convert data tables to 10x format which includes a sparse matrix of gene expression levels, gene names, and barcodes.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># single cell embryo path</span>
<span class="n">path</span> <span class="o">=</span> <span class="s1">&#39;./data/GSE183852_humanHeartFailure/GSE183852_Integrated_Counts.csv.gz&#39;</span>
<span class="c1"># choose columns</span>
<span class="n">usecols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">,</span> <span class="s1">&#39;TWCM-11-103&#39;</span><span class="p">,</span> <span class="s1">&#39;TWCM-13-285&#39;</span><span class="p">]</span>
<span class="c1"># read data with chunks</span>
<span class="n">fr</span> <span class="o">=</span> <span class="n">findRegulators</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
<span class="n">fr</span><span class="o">.</span><span class="n">table_to_10xformat</span><span class="p">(</span>
    <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span>
    <span class="n">transpose</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">chunksize</span><span class="o">=</span><span class="mi">1000</span>
<span class="p">)</span>
</pre></div>
</div>
<p>A similar function <cite>df_to_10xformat</cite> is designed to convert pandas dataframes to 10x format for which the step of reading tables is skipped. Thus, users need to read the data first.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># initialize the object</span>
<span class="n">fr</span> <span class="o">=</span> <span class="n">findSigGenes</span><span class="p">(</span><span class="s1">&#39;./data/scEmbryo/GSE136714/single_cell/&#39;</span><span class="p">)</span>
<span class="c1"># read data</span>
<span class="n">fr</span><span class="o">.</span><span class="n">read_scRNAseq</span><span class="p">()</span>
<span class="c1"># get gene expression matrix</span>
<span class="n">gdf</span> <span class="o">=</span> <span class="n">fr</span><span class="o">.</span><span class="n">get_genedf</span><span class="p">(</span><span class="n">transpose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="c1"># convert data into the 10x format</span>
<span class="n">fr</span><span class="o">.</span><span class="n">df_to_10xformat</span><span class="p">(</span>
        <span class="n">gdf</span><span class="p">,</span>
        <span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;Embryo&#39;</span><span class="p">,</span>
        <span class="p">)</span>
</pre></div>
</div>
</section>
<section id="significant-genes-in-bulk-data">
<h3>Significant genes in bulk data<a class="headerlink" href="#significant-genes-in-bulk-data" title="Permalink to this heading">¶</a></h3>
<p>There are tons of methods used to identify significant genes from bulk omics data, such as DESeq, edgeR, DESeq2, sSeq, and EBSeq. Here, <cite>findSigGenes</cite> not only offers simplest methods like T-test with or without corrections but also integrates DESeq2 and EBSeq to identify significant genes with corresponding P-values and PPDE values.</p>
</section>
<section id="significant-genes-in-single-cells">
<h3>Significant genes in single cells<a class="headerlink" href="#significant-genes-in-single-cells" title="Permalink to this heading">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># instance of the python class</span>
<span class="n">fr</span> <span class="o">=</span> <span class="n">findRegulators</span><span class="p">(</span><span class="s1">&#39;./data/scEmbryo/GSE136714/single_cell/&#39;</span><span class="p">)</span>

<span class="c1"># read RNA-seq data</span>
<span class="n">fr</span><span class="o">.</span><span class="n">read_scRNAseq</span><span class="p">(</span>
        <span class="n">folder</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span>
        <span class="n">rename_cellNames</span><span class="o">=</span><span class="kc">False</span>
        <span class="p">)</span>
<span class="c1"># get gene expression</span>
<span class="n">gdf</span> <span class="o">=</span> <span class="n">fr</span><span class="o">.</span><span class="n">get_genedf</span><span class="p">(</span><span class="n">transpose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># get columns of interest</span>
<span class="n">cols</span> <span class="o">=</span> <span class="n">gdf</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">gdf</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span>
            <span class="s1">&#39;Zygote|2cell|32cell&#39;</span>
            <span class="p">)</span>
        <span class="p">]</span>
<span class="n">dfsel</span> <span class="o">=</span> <span class="n">gdf</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span>
<span class="c1"># find regulators</span>
<span class="n">fr</span> <span class="o">=</span> <span class="n">findRegulators</span><span class="p">(</span><span class="s1">&#39;/nfs/turbo/umms-csriram/daweilin/data/scEmbryo/GSE136714/&#39;</span><span class="p">)</span>
<span class="c1"># dataframe</span>
<span class="n">fr</span><span class="o">.</span><span class="n">df_to_10xformat</span><span class="p">(</span>
        <span class="n">gdf</span><span class="p">,</span>
        <span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;stageEmbryoKSOM&#39;</span><span class="p">,</span>
        <span class="p">)</span>
<span class="n">zygote</span> <span class="o">=</span> <span class="n">dfsel</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">dfsel</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span>
            <span class="s1">&#39;Zygote&#39;</span>
            <span class="p">)</span>
        <span class="p">]</span>
<span class="n">twocell</span> <span class="o">=</span> <span class="n">dfsel</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">dfsel</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;2cell&#39;</span>
            <span class="p">)</span>
        <span class="p">]</span>

<span class="n">bc</span> <span class="o">=</span> <span class="n">dfsel</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">dfsel</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;32cell&#39;</span>
            <span class="p">)</span>
        <span class="p">]</span>
<span class="c1"># get transition of expression between two states</span>
<span class="n">sc1C2C</span> <span class="o">=</span> <span class="n">transition_expression</span><span class="p">(</span><span class="n">dfsel</span><span class="p">[</span><span class="n">zygote</span><span class="p">],</span> <span class="n">dfsel</span><span class="p">[</span><span class="n">twocell</span><span class="p">])</span>
<span class="n">sc2CBC</span> <span class="o">=</span> <span class="n">transition_expression</span><span class="p">(</span><span class="n">dfsel</span><span class="p">[</span><span class="n">twocell</span><span class="p">],</span> <span class="n">dfsel</span><span class="p">[</span><span class="n">bc</span><span class="p">])</span>
<span class="c1"># merge df</span>
<span class="n">mergedf</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">((</span><span class="n">sc1C2C</span><span class="p">,</span> <span class="n">sc2CBC</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="c1"># find regulators</span>
<span class="n">fr</span> <span class="o">=</span> <span class="n">findRegulators</span><span class="p">(</span><span class="s1">&#39;/nfs/turbo/umms-csriram/daweilin/data/scEmbryo/GSE136714/&#39;</span><span class="p">)</span>
<span class="c1"># dataframe</span>
<span class="n">fr</span><span class="o">.</span><span class="n">df_to_10xformat</span><span class="p">(</span>
        <span class="n">mergedf</span><span class="p">,</span>
        <span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;transitionEmbryoKSOM&#39;</span><span class="p">,</span>
        <span class="p">)</span>


<span class="c1"># single cell embryo path</span>
<span class="n">embryoSC_paths</span> <span class="o">=</span> <span class="s1">&#39;/nfs/turbo/umms-csriram/daweilin/data/scEmbryo/GSE136714/GSE136714_raw.xlsx&#39;</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="n">embryoSC_paths</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="c1"># get columns of interest</span>
<span class="n">cols</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span>
            <span class="s1">&#39;Zygote|2cell|32cell&#39;</span>
            <span class="p">)</span>
        <span class="p">]</span>
<span class="n">dfsel</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span>
<span class="c1"># find regulators</span>
<span class="n">fr</span> <span class="o">=</span> <span class="n">findRegulators</span><span class="p">(</span><span class="s1">&#39;/nfs/turbo/umms-csriram/daweilin/data/scEmbryo/GSE136714/&#39;</span><span class="p">)</span>
<span class="c1"># dataframe</span>
<span class="n">fr</span><span class="o">.</span><span class="n">df_to_10xformat</span><span class="p">(</span>
        <span class="n">dfsel</span><span class="p">,</span>
        <span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;stageEmbryoKSOM&#39;</span><span class="p">,</span>
        <span class="p">)</span>
<span class="n">zygote</span> <span class="o">=</span> <span class="n">dfsel</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">dfsel</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span>
            <span class="s1">&#39;Zygote&#39;</span>
            <span class="p">)</span>
        <span class="p">]</span>
<span class="n">twocell</span> <span class="o">=</span> <span class="n">dfsel</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">dfsel</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;2cell&#39;</span>
            <span class="p">)</span>
        <span class="p">]</span>

<span class="n">bc</span> <span class="o">=</span> <span class="n">dfsel</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">dfsel</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;32cell&#39;</span>
            <span class="p">)</span>
        <span class="p">]</span>
<span class="c1"># get transition of expression between two states</span>
<span class="n">sc1C2C</span> <span class="o">=</span> <span class="n">transition_expression</span><span class="p">(</span><span class="n">dfsel</span><span class="p">[</span><span class="n">zygote</span><span class="p">],</span> <span class="n">dfsel</span><span class="p">[</span><span class="n">twocell</span><span class="p">])</span>
<span class="n">sc2CBC</span> <span class="o">=</span> <span class="n">transition_expression</span><span class="p">(</span><span class="n">dfsel</span><span class="p">[</span><span class="n">twocell</span><span class="p">],</span> <span class="n">dfsel</span><span class="p">[</span><span class="n">bc</span><span class="p">])</span>
<span class="c1"># merge df</span>
<span class="n">mergedf</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">((</span><span class="n">sc1C2C</span><span class="p">,</span> <span class="n">sc2CBC</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="c1"># find regulators</span>
<span class="n">fr</span> <span class="o">=</span> <span class="n">findRegulators</span><span class="p">(</span><span class="s1">&#39;/nfs/turbo/umms-csriram/daweilin/data/scEmbryo/GSE136714/&#39;</span><span class="p">)</span>
<span class="c1"># dataframe</span>
<span class="n">fr</span><span class="o">.</span><span class="n">df_to_10xformat</span><span class="p">(</span>
        <span class="n">mergedf</span><span class="p">,</span>
        <span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;transitionEmbryoKSOM&#39;</span><span class="p">,</span>
        <span class="p">)</span>
</pre></div>
</div>
</section>
<section id="classification-significant-genes-with-z-score-and-percentile">
<h3>Classification significant genes with Z-score and percentile<a class="headerlink" href="#classification-significant-genes-with-z-score-and-percentile" title="Permalink to this heading">¶</a></h3>
<p>Again,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="ch">#!/bin/bash</span>
<span class="c1"># input settings</span>
<span class="c1"># path to access your matlab-version cobratoolbox</span>
<span class="nv">COBRA_path</span><span class="o">=</span><span class="s1">&#39;./cobratoolbox/&#39;</span>
<span class="c1"># path to access the metabolic model</span>
<span class="nv">GEM_path</span><span class="o">=</span><span class="s1">&#39;./GEMs/Shen2019.mat&#39;</span>
<span class="c1"># name of the model</span>
<span class="nv">model_name</span><span class="o">=</span><span class="s1">&#39;Recon1&#39;</span>
<span class="c1"># file of the objective candidates including all metabolites from Recon1</span>
<span class="nv">obj_candidate_list_file</span><span class="o">=</span><span class="s1">&#39;./objALL_metabolites_recon1.csv&#39;</span>

<span class="c1"># path to access the significant genes data</span>
<span class="nv">data_dir</span><span class="o">=</span><span class="s1">&#39;./sigGenes/prolif_qui/&#39;</span>
<span class="nv">prefix_name</span><span class="o">=</span><span class="s1">&#39;model&#39;</span><span class="w"> </span><span class="c1"># name of the experiment pls set to &#39;model&#39; for unconstraint models</span>
<span class="nv">medium</span><span class="o">=</span><span class="s1">&#39;DMEMF12&#39;</span><span class="w"> </span><span class="c1"># KSOM for embryos and DMEMF12 for cell culture</span>
<span class="nv">save_root_path</span><span class="o">=</span><span class="s1">&#39;./fluxPrediction/unconstrained_models/pfba/&#39;</span><span class="w"> </span><span class="c1"># path to save predicted fluxes</span>

<span class="c1"># start the simulation of flux predictions</span>
matlab<span class="w"> </span>-nosplash<span class="w"> </span>-noFigureWindows<span class="w"> </span>-r<span class="w"> </span><span class="s2">&quot;multiObj_CBM(~, </span><span class="nv">$COBRA_path</span><span class="s2">, </span><span class="nv">$GEM_path</span><span class="s2">, </span><span class="nv">$model_name</span><span class="s2">, </span><span class="nv">$obj_candidate_list_file</span><span class="s2">, </span><span class="nv">$data_dir</span><span class="s2">, </span><span class="nv">$prefix_name</span><span class="s2">, </span><span class="nv">$medium</span><span class="s2">, </span><span class="nv">$save_root_path</span><span class="s2">)&quot;</span>
</pre></div>
</div>
<p>Similar to what we have done for the models with 52 different metabolites, we have to couple these unconstrained models with omics-basd constrained models to infer metabolic objectives. Notably, the training time and the uncertainty of the coefficients will increase.</p>
</section>
</section>
<section id="consider-more-metabolites-as-single-objectives">
<h2>Consider more metabolites as single objectives<a class="headerlink" href="#consider-more-metabolites-as-single-objectives" title="Permalink to this heading">¶</a></h2>
<p>We have demonstrated that we can infer metabolic objectives based on 52 different metabolites which are mostly considered in Recon1’s biomass objective. However, users might be interested in other metabolites as single objectives. Therefore, this section will guide you to set up the models step by step.</p>
<section id="generate-unconstrained-models-with-more-metabolites">
<h3>Generate unconstrained models with more metabolites<a class="headerlink" href="#generate-unconstrained-models-with-more-metabolites" title="Permalink to this heading">¶</a></h3>
<p>Again, we need to firstly generate the unconstrained models as the variables for SCOOTI.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="ch">#!/bin/bash</span>
<span class="c1"># input settings</span>
<span class="c1"># path to access your matlab-version cobratoolbox</span>
<span class="nv">COBRA_path</span><span class="o">=</span><span class="s1">&#39;./cobratoolbox/&#39;</span>
<span class="c1"># path to access the metabolic model</span>
<span class="nv">GEM_path</span><span class="o">=</span><span class="s1">&#39;./GEMs/Shen2019.mat&#39;</span>
<span class="c1"># name of the model</span>
<span class="nv">model_name</span><span class="o">=</span><span class="s1">&#39;Recon1&#39;</span>
<span class="c1"># file of the objective candidates including all metabolites from Recon1</span>
<span class="nv">obj_candidate_list_file</span><span class="o">=</span><span class="s1">&#39;./objALL_metabolites_recon1.csv&#39;</span>

<span class="c1"># path to access the significant genes data</span>
<span class="nv">data_dir</span><span class="o">=</span><span class="s1">&#39;./sigGenes/prolif_qui/&#39;</span>
<span class="nv">prefix_name</span><span class="o">=</span><span class="s1">&#39;model&#39;</span><span class="w"> </span><span class="c1"># name of the experiment pls set to &#39;model&#39; for unconstraint models</span>
<span class="nv">medium</span><span class="o">=</span><span class="s1">&#39;DMEMF12&#39;</span><span class="w"> </span><span class="c1"># KSOM for embryos and DMEMF12 for cell culture</span>
<span class="nv">save_root_path</span><span class="o">=</span><span class="s1">&#39;./fluxPrediction/unconstrained_models/pfba/&#39;</span><span class="w"> </span><span class="c1"># path to save predicted fluxes</span>

<span class="c1"># start the simulation of flux predictions</span>
matlab<span class="w"> </span>-nosplash<span class="w"> </span>-noFigureWindows<span class="w"> </span>-r<span class="w"> </span><span class="s2">&quot;multiObj_CBM(~, </span><span class="nv">$COBRA_path</span><span class="s2">, </span><span class="nv">$GEM_path</span><span class="s2">, </span><span class="nv">$model_name</span><span class="s2">, </span><span class="nv">$obj_candidate_list_file</span><span class="s2">, </span><span class="nv">$data_dir</span><span class="s2">, </span><span class="nv">$prefix_name</span><span class="s2">, </span><span class="nv">$medium</span><span class="s2">, </span><span class="nv">$save_root_path</span><span class="s2">)&quot;</span>
</pre></div>
</div>
<p>Similar to what we have done for the models with 52 different metabolites, we have to couple these unconstrained models with omics-basd constrained models to infer metabolic objectives. Notably, the training time and the uncertainty of the coefficients will increase.</p>
</section>
<section id="id1">
<h3>Generate unconstrained models with more metabolites<a class="headerlink" href="#id1" title="Permalink to this heading">¶</a></h3>
<p>Again, we need to firstly generate the unconstrained models as the variables for SCOOTI.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="ch">#!/bin/bash</span>
<span class="c1"># input settings</span>
<span class="c1"># path to access your matlab-version cobratoolbox</span>
<span class="nv">COBRA_path</span><span class="o">=</span><span class="s1">&#39;./cobratoolbox/&#39;</span>
<span class="c1"># path to access the metabolic model</span>
<span class="nv">GEM_path</span><span class="o">=</span><span class="s1">&#39;./GEMs/Shen2019.mat&#39;</span>
<span class="c1"># name of the model</span>
<span class="nv">model_name</span><span class="o">=</span><span class="s1">&#39;Recon1&#39;</span>
<span class="c1"># leave it blank if no user-defined objectives</span>
<span class="nv">obj_candidate_list_file</span><span class="o">=</span><span class="s1">&#39;./objALL_metabolites_recon1.csv&#39;</span>
<span class="c1"># objective values</span>
<span class="nv">input_obj_tb</span><span class="o">=</span><span class="s1">&#39;&#39;</span>

<span class="c1"># parameter settings</span>
<span class="nv">DFA_kappa</span><span class="o">=</span>-1
<span class="nv">CFR_kappa</span><span class="o">=</span><span class="m">0</span>.1
<span class="nv">CFR_rho</span><span class="o">=</span><span class="m">10</span>
<span class="nv">paraLen</span><span class="o">=</span><span class="m">1</span><span class="w"> </span><span class="c1"># how many kappa/rho used for scanning</span>
<span class="nv">random_para</span><span class="o">=</span><span class="m">0</span><span class="w"> </span><span class="c1"># bool, 1 to enable random sampling</span>
<span class="nv">init_objective</span><span class="o">=</span><span class="m">1</span><span class="w"> </span><span class="c1"># 1 for none, 2 for biomass objective</span>
<span class="nv">genekoflag</span><span class="o">=</span><span class="m">0</span><span class="w"> </span><span class="c1"># bool</span>
<span class="nv">rxnkoflag</span><span class="o">=</span><span class="m">0</span><span class="w"> </span><span class="c1"># bool</span>
<span class="nv">FVAflag</span><span class="o">=</span><span class="m">0</span><span class="w"> </span><span class="c1"># bool</span>
<span class="nv">pfba</span><span class="o">=</span><span class="m">1</span><span class="w"> </span><span class="c1"># 0 for fba and 1 for pfba (minimize sum of fluxes)</span>
<span class="nv">medium_perturbation</span><span class="o">=</span><span class="m">0</span><span class="w"> </span><span class="c1"># 1 for depletion or excess of metabolites in medium</span>
<span class="nv">pairwise_CFR_model</span><span class="o">=</span><span class="m">0</span>
<span class="nv">algorithm</span><span class="o">=</span><span class="s1">&#39;iMAT&#39;</span>

<span class="c1"># path to access the significant genes data</span>
<span class="nv">data_dir</span><span class="o">=</span><span class="s1">&#39;./sigGenes/prolif_qui/&#39;</span>
<span class="nv">prefix_name</span><span class="o">=</span><span class="s1">&#39;model&#39;</span><span class="w"> </span><span class="c1"># name of the experiment pls set to &#39;model&#39; for unconstraint models</span>
<span class="nv">medium</span><span class="o">=</span><span class="s1">&#39;DMEMF12&#39;</span><span class="w"> </span><span class="c1"># KSOM for embryos and DMEMF12 for cell culture</span>
<span class="nv">late_stage</span><span class="o">=</span><span class="s1">&#39;upgenes&#39;</span><span class="w"> </span><span class="c1"># suffix of the file names of significant up-genes</span>
<span class="nv">early_stage</span><span class="o">=</span><span class="s1">&#39;dwgenes&#39;</span><span class="w"> </span><span class="c1"># suffix of the file names of significant down-genes</span>
<span class="nv">simulation</span><span class="o">=</span><span class="s1">&#39;CFR&#39;</span><span class="w"> </span><span class="c1"># CFR for transcriptomics and proteomics; DFA for metabolomics</span>
<span class="nv">constraint</span><span class="o">=</span><span class="m">1</span><span class="w"> </span><span class="c1"># apply constraints to the model</span>
<span class="nv">save_root_path</span><span class="o">=</span><span class="s1">&#39;./fluxPrediction/unconstrained_models/pfba/&#39;</span><span class="w"> </span><span class="c1"># path to save predicted fluxes</span>
<span class="nv">CFR_model_path</span><span class="o">=</span><span class="s1">&#39;&#39;</span>


<span class="nv">run</span><span class="o">=</span><span class="m">1</span>
matlab<span class="w"> </span>-nosplash<span class="w"> </span>-noFigureWindows<span class="w"> </span>-r<span class="w"> </span><span class="s2">&quot;multiObj_CBM(</span><span class="nv">$run</span><span class="s2">, </span><span class="nv">$DFA_kappa</span><span class="s2">, </span><span class="nv">$CFR_kappa</span><span class="s2">, </span><span class="nv">$CFR_rho</span><span class="s2">, </span><span class="nv">$COBRA_path</span><span class="s2">,</span><span class="nv">$GEM_path</span><span class="s2">, </span><span class="nv">$model_name</span><span class="s2">, </span><span class="nv">$obj_candidate_list_file</span><span class="s2">, </span><span class="nv">$input_obj_tb</span><span class="s2">, </span><span class="nv">$paraLen</span><span class="s2">, </span><span class="nv">$random_para</span><span class="s2">, </span><span class="nv">$init_objective</span><span class="s2">, </span><span class="nv">$genekoflag</span><span class="s2">, </span><span class="nv">$rxnkoflag</span><span class="s2">, </span><span class="nv">$FVAflag</span><span class="s2">, </span><span class="nv">$pfba</span><span class="s2">, </span><span class="nv">$medium_perturbation</span><span class="s2">, </span><span class="nv">$data_dir</span><span class="s2">, </span><span class="nv">$prefix_name</span><span class="s2">, </span><span class="nv">$medium</span><span class="s2">, </span><span class="nv">$late_stage</span><span class="s2">, </span><span class="nv">$early_stage</span><span class="s2">, </span><span class="nv">$simulation</span><span class="s2">, </span><span class="nv">$constraint</span><span class="s2">, </span><span class="nv">$save_root_path</span><span class="s2">)&quot;</span>
</pre></div>
</div>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">SCOOTI</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quick Start</a></li>
<li class="toctree-l1"><a class="reference internal" href="scootisettings.html">SCOOTI Settings</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Advanced usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="#identifying-up-and-down-regulated-genes">Identifying up- and down-regulated genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="#consider-more-metabolites-as-single-objectives">Consider more metabolites as single objectives</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="scootisettings.html" title="previous chapter">SCOOTI Settings</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2023, Chandrasekaran Lab, University of Michigan.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.3.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.13</a>
      
      |
      <a href="_sources/advancedusage.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>