
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Welcome to SCOOBI’s documentation! &#8212; SCOOTI 0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="module-SCOOTI.metObjAnalyzer">
<span id="welcome-to-scoobi-s-documentation"></span><h1>Welcome to SCOOBI’s documentation!<a class="headerlink" href="#module-SCOOTI.metObjAnalyzer" title="Permalink to this heading">¶</a></h1>
<section id="metobjanalyzer-py">
<h2>metObjAnalyzer.py<a class="headerlink" href="#metobjanalyzer-py" title="Permalink to this heading">¶</a></h2>
<p>Object-oriented classes that serve as a user interface to run analysis/simulations</p>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">SCOOTI.metObjAnalyzer.</span></span><span class="sig-name descname"><span class="pre">metObjAnalyzer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">flux_paths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coef_paths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_root_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">GEM_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">uncon_model_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_map</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samplingFlux_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sel_para</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'KSOM'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer" title="Permalink to this definition">¶</a></dt>
<dd><p>Analysis of optimized fluxes and metabolic objectives.</p>
<p>The object consists of four parts of analysis.
1) Optimized flux:</p>
<blockquote>
<div><p>suggests the best parameters and visualization.</p>
</div></blockquote>
<ol class="arabic simple" start="2">
<li><dl class="simple">
<dt>Metabolic objective:</dt><dd><p>compare to different samples and allocations.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Tradeoffs and traits:</dt><dd><p>identify anticorrelated metabolites and metabolic traits.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Flux reconstruction:</dt><dd><p>reconstruct flux prediction with inferred coefficients.</p>
</dd>
</dl>
</li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>flux_paths</strong> (<em>{dictionary}</em><em>,</em>) – Paths to access the directories with flux prediction data.
Keys should be the labels (sample names) associated with the directories.</p></li>
<li><p><strong>coef_paths</strong> (<em>{dictionary}</em><em>,</em>) – Paths to access the files (.csv) with inferred metabolic objectives.
Keys should be the labels (sample names) associated with the files.</p></li>
<li><p><strong>save_root_path</strong> (<em>{str}</em><em>,</em>) – Path to save the outputs (including files and figures).
Absolute path is recommended.</p></li>
<li><p><strong>GEM_path</strong> (<em>{str}</em><em>,</em>) – Paths to access the genome-scale metabolic models (.mat files).
Only Shen et al., Recon2.2, and Recon3D are applicable for this version.</p></li>
<li><p><strong>uncon_model_path</strong> (<em>{str}</em><em>,</em>) – Paths to access the directories with flux predictions of single-objective models.</p></li>
<li><p><strong>col_map</strong> (<em>{dictionary}</em><em>,</em>) – Serialize unique column names (samples) and corresponding preferred names.</p></li>
<li><p><strong>label_func</strong> (<em>{function}</em><em>,</em>) – A function to convert column names into unique sample names.</p></li>
<li><p><strong>samplingFlux_path</strong> (<em>{str}</em><em>,</em>) – Paths to access the directories with flux predictions
models with randomly sampled objective coefficients.</p></li>
<li><p><strong>sel_para</strong> (<em>{str}</em><em>,</em>) – A parameter combination selected for CFR (e.g. k0.1_r0.01)</p></li>
<li><p><strong>prefix</strong> (<em>{str}</em><em>,</em>) – Project name (e.g. scEmbryo)</p></li>
<li><p><strong>medium</strong> (<em>{string}</em><em>, </em><em>options=</em><em>[</em><em>'KSOM'</em><em>, </em><em>'DMEMF12'</em><em>]</em>) – medium used to model the fluxes with FBA</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.flux_df">
<span class="sig-name descname"><span class="pre">flux_df</span></span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.flux_df" title="Permalink to this definition">¶</a></dt>
<dd><p>Flux data predicted with constrained models without objectives</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>{pandas.DataFrame},</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.coef_df">
<span class="sig-name descname"><span class="pre">coef_df</span></span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.coef_df" title="Permalink to this definition">¶</a></dt>
<dd><p>Coefficients of metabolic objectives</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>{pandas.DataFrame},</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.labels">
<span class="sig-name descname"><span class="pre">labels</span></span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.labels" title="Permalink to this definition">¶</a></dt>
<dd><p>label for each column of flux or coefficient data</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>{pandas.DataFrame},</p>
</dd>
</dl>
</dd></dl>

<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">SCOOTI.metObjAnalyzer</span> <span class="kn">import</span> <span class="n">metObjAnalyzer</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># set up a function to replace column names</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">def</span> <span class="nf">label_func</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">({</span><span class="s1">&#39;1C&#39;</span><span class="p">:</span><span class="s1">&#39;sc1C2C&#39;</span><span class="p">,</span> <span class="s1">&#39;2C&#39;</span><span class="p">:</span><span class="s1">&#39;sc2CBC&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># initiate the object</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">moa</span> <span class="o">=</span> <span class="n">metObjAnalyzer</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">flux_paths</span><span class="o">=</span><span class="p">{</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="s1">&#39;sc1C2C&#39;</span><span class="p">:</span><span class="s1">&#39;./fluxPrediction/sc1C2C/paraScan/&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="s1">&#39;sc2CBC&#39;</span><span class="p">:</span><span class="s1">&#39;./fluxPrediction/sc2CBC/paraScan/&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>                <span class="p">},</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">coef_paths</span><span class="o">=</span><span class="p">{</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="s1">&#39;sc1C2C&#39;</span><span class="p">:</span><span class="sa">f</span><span class="s1">&#39;./regression_models/scEmbryo_paraScan/flux_sl_sc1C2C_input_norm_outcome_nonorm_k0.1_r0.01.csv&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="s1">&#39;sc2CBC&#39;</span><span class="p">:</span><span class="sa">f</span><span class="s1">&#39;./regression_models/scEmbryo_paraScan/flux_sl_sc2CBC_input_norm_outcome_nonorm_k0.1_r0.01.csv&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">},</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">save_root_path</span><span class="o">=</span><span class="s1">&#39;/home/daweilin/StemCell/Project_mESC_JinZhang/regressor_results_new/&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">GEM_path</span><span class="o">=</span><span class="s1">&#39;./models/Recon1.mat&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">uncon_model_path</span><span class="o">=</span><span class="s1">&#39;./fluxPrediction/unconstrained_models/pfba/KSOM/&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">col_map</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;sc1C2C&#39;</span><span class="p">:</span><span class="s1">&#39;sc1C2C&#39;</span><span class="p">,</span> <span class="s1">&#39;sc2CBC&#39;</span><span class="p">:</span><span class="s1">&#39;sc2CBC&#39;</span><span class="p">},</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">label_func</span><span class="o">=</span><span class="n">label_func</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">samplingFlux_path</span><span class="o">=</span><span class="s1">&#39;./fluxPrediction/RandomObjCoef/embryo_ksom_50percent/&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">sel_para</span><span class="o">=</span><span class="s1">&#39;k0.1_r0.01&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;scEmbryo&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">medium</span><span class="o">=</span><span class="s1">&#39;KSOM&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Analyze fluxes and look for the best parameter combinations</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">moa</span><span class="o">.</span><span class="n">fluxAnalysis</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">kappa_arr</span><span class="o">=</span><span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">rho_arr</span><span class="o">=</span><span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># read coefficients</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">moa</span><span class="o">.</span><span class="n">get_coef</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># analyze coefficients</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coefAnalysis</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">clustering</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">entropy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">distance</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">umap_para</span><span class="o">=</span><span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">50</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="p">)</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.coefAnalysis">
<span class="sig-name descname"><span class="pre">coefAnalysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">norm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unknown_clustering</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clustering</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entropy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umap_para</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[5,</span> <span class="pre">50]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recluster_min_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'average'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">get_randomObj_coef</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">th</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.coefAnalysis" title="Permalink to this definition">¶</a></dt>
<dd><p>Analysis of inferred metabolic objectives.</p>
<p>The coefficients of metabolic objectives were analyzed by
1) comparing to different samples,
2) dimension reduction methods for metabolic objectives,
3) objective entropy (degree of order),
4) distances from the biomass objective to the inferred objectives</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>norm</strong> (<em>{bool}</em><em>, </em><em>default=True</em>) – normalize the coefficients by the sum of all coefficients of a sample.</p></li>
<li><p><strong>clustering</strong> (<em>{bool}</em><em>, </em><em>default=True</em>) – enable dimension reduction methods or clustering methods for metabolic objectives.
UMAP is the default method in the class, but it is allowed to be swapped by
PCA, tSNE, and PHATE.</p></li>
<li><p><strong>entropy</strong> (<em>{bool}</em><em>, </em><em>default=True</em>) – enable the calculations of entropy based on the coefficients of metabolic objectives
which follows the equation derived from <cite>p(log(p))</cite>. <cite>p</cite> is converted from</p></li>
<li><p><strong>distance</strong> (<em>{bool}</em><em>, </em><em>default=True</em>) – enable dimension reduction methods or clustering methods for metabolic objectives</p></li>
<li><p><strong>umap_para</strong> (<em>{array-like}</em><em>, </em><em>default=</em><em>[</em><em>5</em><em>, </em><em>10</em><em>]</em>) – the first number of the array stands for the number of neighbors and
the second number of the array is the total dimension of UMAP</p></li>
<li><p><strong>recluster_min_size</strong> (<em>{int}</em><em>, </em><em>default=10</em>) – the parameter for HDBSCAN to cluster the UMAP results</p></li>
<li><p><strong>method</strong> (<em>{str}</em><em>, </em><em>default='average'</em>) – method used to cluster samples with scipy.cluster.hierarchy.linkage.
Options include ‘average’, ‘single’, ‘complete’, ‘weighted’, ‘centroid’.
For more details, one can visit the official document.
<a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.linkage.html">https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.linkage.html</a></p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>get_randomObj_coef<span class="classifier">{bool}, default=False</span></dt><dd><p>enable randomly sampling coefficients of selected metabolites based on how many portion
of cells choose a metabolite as a feature. The coefficients will be saved into a .csv file</p>
</dd>
<dt>th<span class="classifier">{float}, default=50</span></dt><dd><p>the threshold to select metabolites as features. the number represents the percentage of
cells that choose the metabolites; thus, it ranges from [0, 100]</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>self</strong> – Returns self.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>object</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The figures of the analysis will be saved in the path given by <cite>self.save_root_path</cite>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.fluxAnalysis">
<span class="sig-name descname"><span class="pre">fluxAnalysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kappa_arr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[10,</span> <span class="pre">1,</span> <span class="pre">0.1,</span> <span class="pre">0.01,</span> <span class="pre">0.001]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rho_arr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[10,</span> <span class="pre">1,</span> <span class="pre">0.1,</span> <span class="pre">0.01,</span> <span class="pre">0.001]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umap_para</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[5,</span> <span class="pre">10]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.fluxAnalysis" title="Permalink to this definition">¶</a></dt>
<dd><p>Cluster fluxes of UMAP with HDBSCAN and output clustering perfomance.</p>
<p>Cell-specific constrained models without objective functions
1) The model predicts the fluxes of ideal optimality
2) We aim to model fluxes that can separate cell types</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>kappa_arr</strong> (<em>{array-like}</em><em> of </em><em>shape</em><em> (</em><em>n_parameters</em><em>,</em><em>)</em><em>, </em><em>default=</em><em>[</em><em>10</em><em>, </em><em>1</em><em>, </em><em>0.1</em><em>, </em><em>0.01</em><em>, </em><em>0.001</em><em>]</em>) – parameter of interest.</p></li>
<li><p><strong>rho_arr</strong> (<em>{array-like}</em><em> of </em><em>shape</em><em> (</em><em>n_parameters</em><em>,</em><em>)</em><em>, </em><em>default=</em><em>[</em><em>10</em><em>, </em><em>1</em><em>, </em><em>0.1</em><em>, </em><em>0.01</em><em>, </em><em>0.001</em><em>]</em>) – parameter of interest.</p></li>
<li><p><strong>umap_para</strong> (<em>{array-like}</em><em>, </em><em>default=</em><em>[</em><em>5</em><em>, </em><em>10</em><em>]</em>) – the first number of the array stands for the number of neighbors and
the second number of the array is the total dimension of UMAP</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>self</strong> (<em>object</em>) – Returns self.</p></li>
<li><p><strong>flux_df</strong> (<em>{pandas.DataFrame} of shape (n_reactions, n_samples), optional</em>) – The data table of flux predictions generated from either CFR or DFA.
Only provided if <cite>self.sel_para</cite> is given in the class attributes.</p></li>
<li><p><strong>scores</strong> (<em>{pandas.DataFrame} of shape (n_parameters, 4), optional</em>) – The table of clustering evaluations which comes with coverage rate of HDBSCAN,
rand index, mutual information, and silhouette scores.
It is only provided if <cite>self.sel_para</cite> is not given in the class attributes.
A .csv file with the evaluations is output at the same time.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.fluxRecon">
<span class="sig-name descname"><span class="pre">fluxRecon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">clustering</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">validation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.fluxRecon" title="Permalink to this definition">¶</a></dt>
<dd><p>Flux reconstruction</p>
<p>Predict the fluxes by  coefficients*unconstrained models</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>clustering</strong> (<em>{bool}</em><em>, </em><em>default=True</em>) – enable dimension reduction methods or clustering methods for metabolic objectives.
UMAP is the default method in the class, but it is allowed to be swapped by
PCA, tSNE, and PHATE.</p></li>
<li><p><strong>validation</strong> (<em>{bool}</em><em>, </em><em>default=False</em>) – enable the evaluation of regression models with R2, Pearson and Spearman correlations.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>self</strong> (<em>object</em>) – Returns self.</p></li>
<li><p><strong>res</strong> (<em>{pandas.DataFrame}, optional</em>) – a table of evaluation results. Only provided if <cite>validation</cite> is set true.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.functionalAnalysis">
<span class="sig-name descname"><span class="pre">functionalAnalysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cellTypes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sheetnames</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expression_tb_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gesa_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pluri_gene_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sigGenes_root_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.functionalAnalysis" title="Permalink to this definition">¶</a></dt>
<dd><p>Functional analysis overlaying with single-cell data</p>
<p>Refer to functional genes from MSigDB for hypothesis validation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cellTypes</strong> (<em>{string array}</em><em>,</em>) – names of the labels used for analysis. The array should contain unique elements.</p></li>
<li><p><strong>sheetnames</strong> (<em>{string array}</em><em>,</em>) – sheetnames to access the data of significant genes corresponding to <cite>cellTypes</cite>.</p></li>
<li><p><strong>expression_tb_path</strong> (<em>{string}</em><em>,</em>) – path to access the file of gene expression data.</p></li>
<li><p><strong>gesa_name</strong> (<em>{string}</em><em>,</em>) – directory to access all the functional genes.</p></li>
<li><p><strong>pluri_gene_path</strong> (<em>{string}</em><em>,</em>) – path to access the directories where save different types of functional genes.</p></li>
<li><p><strong>sigGenes_root_path</strong> (<em>{string}</em><em>,</em>) – path to access the files of significant genes.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>self</strong> – Returns self.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>object</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This method is only applicable for specific structure of directories and data in Version 0.0.1.</p>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># umap clustering for cluster separation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cf</span> <span class="o">=</span> <span class="n">clustering_func</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>            <span class="n">coef_sel</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="n">coef_sel</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
<span class="gp">&gt;&gt;&gt; </span>            <span class="s1">&#39;./result_figs/&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>            <span class="sa">f</span><span class="s1">&#39;scEmbryo_coef_for_gsea&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>            <span class="n">mets_category</span><span class="p">(),</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># get umap components</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">umaps</span> <span class="o">=</span> <span class="n">cf</span><span class="o">.</span><span class="n">reduction_scatter</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">coef_sel</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">continuous</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">func</span><span class="o">=</span><span class="s1">&#39;UMAP&#39;</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># cell types</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cells</span> <span class="o">=</span> <span class="n">umaps</span><span class="p">[</span><span class="s1">&#39;Cell type&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">exptb</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s1">&#39;/home/daweilin/StemCell/Project_mESC_JinZhang/validation_data/GSE136714/GSE136714_raw.xlsx&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># cluster the umap results</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">recluster</span> <span class="o">=</span> <span class="n">cf</span><span class="o">.</span><span class="n">reclustering</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">umaps</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;UMAP1&#39;</span><span class="p">,</span> <span class="s1">&#39;UMAP2&#39;</span><span class="p">],</span> <span class="n">min_size</span><span class="o">=</span><span class="mi">5</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># functional analysis</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">GESA_analyais</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">umaps</span><span class="p">,</span> <span class="n">recluster</span><span class="p">,</span> <span class="s1">&#39;Cell type&#39;</span><span class="p">,</span> <span class="n">exptb</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">gesa_name</span><span class="o">=</span><span class="s1">&#39;pluripotency&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">cellTypes</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;1C2C&#39;</span><span class="p">,</span> <span class="s1">&#39;2CBC&#39;</span><span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">sheetnames</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;2C&#39;</span><span class="p">,</span> <span class="s1">&#39;BC&#39;</span><span class="p">],</span> <span class="n">reduction_func</span><span class="o">=</span><span class="s1">&#39;UMAP&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">pluri_gene_path</span><span class="o">=</span><span class="s1">&#39;./mcdb/&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">sigGenes_root_path</span><span class="o">=</span><span class="s1">&#39;./data/scEmbryo/GSE136714/sigGenes/&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">save_root_path</span><span class="o">=</span><span class="s1">&#39;./result_figs/&#39;</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.getSpecificSamples">
<span class="sig-name descname"><span class="pre">getSpecificSamples</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sel_label_arr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">update_attr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.getSpecificSamples" title="Permalink to this definition">¶</a></dt>
<dd><p>Slice table with columns associated with correspondingly selected labels</p>
<p>Users select specific columns via given labels to chop the data table.
Since the column names are not the same as the label names,
selected labels will compare with the entire array of labels to index the columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>labels</strong> (<em>{array-like}</em><em> of </em><em>shape</em><em> (</em><em>n_columns</em><em>,</em><em>)</em>) – name array that should has the same length of columns</p></li>
<li><p><strong>sel_label_arr</strong> (<em>{array-like}</em>) – names chosen from labels that should be unique</p></li>
<li><p><strong>update_attr</strong> (<em>{bool}</em><em>, </em><em>default=False</em>) – update class attributes of flux/coefficients dataframe and labels
if update_attr is set true</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>self</strong> – returns self</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>object</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.get_coef">
<span class="sig-name descname"><span class="pre">get_coef</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.get_coef" title="Permalink to this definition">¶</a></dt>
<dd><p>Load inferred metabolic objectives.</p>
<p>The coefficients of metabolic objectives were obtained by
the unconstrained models with single objectives regressiing on
the condition/cell-specific constrained models</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>kappa_arr</strong> (<em>{array-like}</em><em> of </em><em>shape</em><em> (</em><em>n_parameters</em><em>,</em><em>)</em><em>, </em><em>default=</em><em>[</em><em>10</em><em>, </em><em>1</em><em>, </em><em>0.1</em><em>, </em><em>0.01</em><em>, </em><em>0.001</em><em>]</em>) – parameter of interest.</p></li>
<li><p><strong>rho_arr</strong> (<em>{array-like}</em><em> of </em><em>shape</em><em> (</em><em>n_parameters</em><em>,</em><em>)</em><em>, </em><em>default=</em><em>[</em><em>10</em><em>, </em><em>1</em><em>, </em><em>0.1</em><em>, </em><em>0.01</em><em>, </em><em>0.001</em><em>]</em>) – parameter of interest.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>self</strong> – Returns self.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>object</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.get_flux">
<span class="sig-name descname"><span class="pre">get_flux</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kappa</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rho</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.get_flux" title="Permalink to this definition">¶</a></dt>
<dd><p>Load inferred metabolic objectives.</p>
<p>The coefficients of metabolic objectives were obtained by
the unconstrained models with single objectives regressiing on
the condition/cell-specific constrained models</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>kappa_arr</strong> (<em>{array-like}</em><em> of </em><em>shape</em><em> (</em><em>n_parameters</em><em>,</em><em>)</em><em>, </em><em>default=</em><em>[</em><em>10</em><em>, </em><em>1</em><em>, </em><em>0.1</em><em>, </em><em>0.01</em><em>, </em><em>0.001</em><em>]</em>) – parameter of interest.</p></li>
<li><p><strong>rho_arr</strong> (<em>{array-like}</em><em> of </em><em>shape</em><em> (</em><em>n_parameters</em><em>,</em><em>)</em><em>, </em><em>default=</em><em>[</em><em>10</em><em>, </em><em>1</em><em>, </em><em>0.1</em><em>, </em><em>0.01</em><em>, </em><em>0.001</em><em>]</em>) – parameter of interest.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>self</strong> – Returns self.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>object</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.tradeoff_analysis">
<span class="sig-name descname"><span class="pre">tradeoff_analysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tri_mets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['gh',</span> <span class="pre">'chsterol',</span> <span class="pre">'glutathione']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pareto_mets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['gh']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">triangle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pareto3D</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">traitAnalysis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_trait_func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_trait_func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.tradeoff_analysis" title="Permalink to this definition">¶</a></dt>
<dd><p>Analysis of inferred metabolic objectives.</p>
<p>The coefficients of metabolic objectives were analyzed by
1) correlation of coefficients: a pair of metabolites that compete for allocations,
2) triangle plot: tradeoffs and relationships among three metabolites,
3) pareto analysis: the realistic allocations compared to the optimal solutions in 2D or 3D,
4) trait analysis: PCA-reduced allocations used to identify metabolic traits</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tri_mets</strong> (<em>{string array}</em><em>, </em><em>default=</em><em>[</em><em>'gh'</em><em>, </em><em>'chsterol'</em><em>, </em><em>'glutathione'</em><em>]</em>) – Input of metabolites for the triangle plot. The length of the input array has to be 3.
<cite>glutathione</cite>, the label, is only applicable when using Recon1 or Shen model which merges
the coefficients of oxidized and reduced glutathione (<cite>gthox</cite> and <cite>gthrd</cite>).
Only useful when the parameter <cite>triangle</cite> is set true.</p></li>
<li><p><strong>pareto_mets</strong> (<em>{string array}</em><em>, </em><em>default=</em><em>[</em><em>'gh'</em><em>]</em>) – Input of metabolites for the plots of i) scatters of relationships or ii) Pareto fronts between pairs of metabolites.
The length of the input array has to be at least 1.</p></li>
<li><p><strong>triangle</strong> (<em>{bool}</em><em>, </em><em>default=False</em>) – enable triangle plots of allocations among the three metabolites defined in <cite>tri_mets</cite>.</p></li>
<li><p><strong>pareto3D</strong> (<em>{bool}</em><em>, </em><em>default=False</em>) – enable Pareto analysis among three metabolites provided in the array of <cite>pareto_mets</cite>
which includes the calculation of Euclidean distance from the data points to the surface.</p></li>
<li><p><strong>traitAnalysis</strong> (<em>{bool}</em><em>, </em><em>default=False</em>) – enable PCA-based analysis of allocations in high dimensional objective coefficients.
The vertices of traits represent the extreme traits of the metabolic objectives
based on the mechanistic limitation.</p></li>
<li><p><strong>sample_trait_func</strong> (<em>{bool}</em><em>, </em><em>default=False</em>) – additional function that gets omics-based data points closer to traits on vertices</p></li>
<li><p><strong>control_trait_func</strong> (<em>{bool}</em><em>, </em><em>default=False</em>) – additional function that gets mechanistic data points closet to traits on vertices</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>self</strong> – Returns self.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>object</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<span class="target" id="module-SCOOTI.regressionAnalyzer"></span><section id="regressionanalyzer-py">
<h2>regressionAnalyzer.py<a class="headerlink" href="#regressionanalyzer-py" title="Permalink to this heading">¶</a></h2>
<p>A collection of methods for analysis of metabolic objectives and flux predictions</p>
</section>
<dl class="py function">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.EXflux_validation">
<span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">EXflux_validation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">match_map_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ffm_cp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multiObj_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">uncon_res</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_replace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overall_correlation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.EXflux_validation" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>(</strong><strong>pandas.DataFrame</strong><strong>)</strong> (<em>multiObj_df</em>) – </p></li>
<li><p><strong>(</strong><strong>pandas.DataFrame</strong><strong>)</strong> – </p></li>
<li><p><strong>(</strong><strong>pandas.DataFrame</strong><strong>)</strong> – </p></li>
<li><p><strong>(</strong><strong>bool</strong><strong>)</strong> (<em>col_replace</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>(dictionary)</em></p></li>
<li><p><em>{</em></p></li>
<li><p><strong>‘correlation’</strong> (<em>valid_corr_df (pandas.DataFrame)</em>)</p></li>
<li><p><strong>’predicted_fluxes’</strong> (<em>flux_valid (pandas.DataFrame)</em>)</p></li>
<li><p><strong>’stacking_correlation’</strong> (<em>corr_all (float)</em>)</p></li>
<li><p><strong>’pvalues’</strong> (<em>pvs (numpy.array)</em>)</p></li>
<li><p><strong>’coefficients’</strong> (<em>coefs (numpy.array)</em>)</p></li>
<li><p><strong>’CI_lb’</strong> (<em>CI_lb (numpy.array)</em>)</p></li>
<li><p><strong>’CI_ub’</strong> (<em>CI_ub (numpy.array)</em>)</p></li>
<li><p><em>}</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.GESA_analyais">
<span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">GESA_analyais</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pc_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reclusters</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exptb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gesa_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'trophectoderm'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cellTypes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['1C2C',</span> <span class="pre">'2CBC']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sheetnames</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['2C',</span> <span class="pre">'BC']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reduction_func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'UMAP'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pluri_gene_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/home/daweilin/StemCell/Project_mESC_JinZhang/mcdb/'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sigGenes_root_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/nfs/turbo/umms-csriram/daweilin/data/scEmbryo/GSE136714/old_sigGenes/'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_root_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/home/daweilin/StemCell/Project_mESC_JinZhang/regressor_results_new/'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.GESA_analyais" title="Permalink to this definition">¶</a></dt>
<dd><section id="description">
<h2>Description<a class="headerlink" href="#description" title="Permalink to this heading">¶</a></h2>
<p>1. Determine if significant genes in each data point can be categoried into
a certain GESA function by hypergeometic tests aka functional analysis
2. Dimension reudction plots overlaies p-values of functional analysis
3. Compare p-values in clusters (reclustered by HDBSCAN)</p>
<dl class="field-list simple">
<dt class="field-odd">param pc_df (pandas.DataFrame)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type pc_df (pandas.DataFrame)<span class="colon">:</span></dt>
<dd class="field-even"><p>dimension reduction results</p>
</dd>
<dt class="field-odd">param reclusters (pandas.DataFrame)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type reclusters (pandas.DataFrame)<span class="colon">:</span></dt>
<dd class="field-even"><p>reclusters of dimension reduction results</p>
</dd>
<dt class="field-odd">param label_col (str)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type label_col (str)<span class="colon">:</span></dt>
<dd class="field-even"><p>column name to access a list of names corresponding to the pc data</p>
</dd>
<dt class="field-odd">param exptb (pandas.DataFrame)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type exptb (pandas.DataFrame)<span class="colon">:</span></dt>
<dd class="field-even"><p>gene expression data</p>
</dd>
<dt class="field-odd">param gesa_name (str)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type gesa_name (str)<span class="colon">:</span></dt>
<dd class="field-even"><p>folder name of the data (embryoDevelopment/pluripotency/proliferation)</p>
</dd>
<dt class="field-odd">param cellTypes (list)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type cellTypes (list)<span class="colon">:</span></dt>
<dd class="field-even"><p>a list of names to access the folders that save significant gene data</p>
</dd>
<dt class="field-odd">param sheetnames (list)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type sheetnames (list)<span class="colon">:</span></dt>
<dd class="field-even"><p>a list of names to access the excel sheets/csv that save sig genes</p>
</dd>
<dt class="field-odd">param reduction_func (str)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type reduction_func (str)<span class="colon">:</span></dt>
<dd class="field-even"><p>name of the function of reduction methods used to access columns</p>
</dd>
<dt class="field-odd">param sigGenes_root_path (str)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type sigGenes_root_path (str)<span class="colon">:</span></dt>
<dd class="field-even"><p>path to access the significant genes</p>
</dd>
<dt class="field-odd">param save_root_path (str)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type save_root_path (str)<span class="colon">:</span></dt>
<dd class="field-even"><p>path to save figures</p>
</dd>
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.Objflux_validation">
<span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">Objflux_validation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">match_map_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ffm_cp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multiObj_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_replace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overall_correlation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.Objflux_validation" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>(</strong><strong>pandas.DataFrame</strong><strong>)</strong> (<em>multiObj_df</em>) – </p></li>
<li><p><strong>(</strong><strong>pandas.DataFrame</strong><strong>)</strong> – </p></li>
<li><p><strong>(</strong><strong>pandas.DataFrame</strong><strong>)</strong> – </p></li>
<li><p><strong>(</strong><strong>bool</strong><strong>)</strong> (<em>col_replace</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>(dictionary)</em></p></li>
<li><p><em>{</em></p></li>
<li><p><strong>‘correlation’</strong> (<em>valid_corr_df (pandas.DataFrame)</em>)</p></li>
<li><p><strong>’predicted_fluxes’</strong> (<em>flux_valid (pandas.DataFrame)</em>)</p></li>
<li><p><strong>’stacking_correlation’</strong> (<em>corr_all (float)</em>)</p></li>
<li><p><strong>’pvalues’</strong> (<em>pvs (numpy.array)</em>)</p></li>
<li><p><strong>’coefficients’</strong> (<em>coefs (numpy.array)</em>)</p></li>
<li><p><strong>’CI_lb’</strong> (<em>CI_lb (numpy.array)</em>)</p></li>
<li><p><strong>’CI_ub’</strong> (<em>CI_ub (numpy.array)</em>)</p></li>
<li><p><em>}</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.allocation_plot">
<span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">allocation_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">coef_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cellType</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.allocation_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>The allocation of each metabolite in inferred metabolic objectives.</p>
<p>The allocation of coefficients will be represented by
c/Sigma(c) if norm is set true</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>coef_df</strong> (<em>{pandas.DataFrame}</em><em> of </em><em>shape</em><em> (</em><em>n_metabolites</em><em>, </em><em>n_samples</em><em>)</em><em>,</em>) – the coefficients of metabolic objectives for each samples</p></li>
<li><p><strong>labels</strong> (<em>{array-like}</em><em> of </em><em>shape</em><em> (</em><em>n_samples</em><em>,</em><em>)</em><em>,</em>) – the labels of samples/columns</p></li>
<li><p><strong>cellType</strong> (<em>{string}</em><em>,</em>) – the label selected to show allocations</p></li>
<li><p><strong>prefix</strong> (<em>{string}</em><em>, </em><em>default=''</em>) – experiment names for saving files</p></li>
<li><p><strong>norm</strong> (<em>{bool}</em><em>, </em><em>default=True</em>) – normalize coefficients by the sum of all the coefficients</p></li>
<li><p><strong>cutoff</strong> (<em>{float}</em><em>, </em><em>default=0.0</em>) – thresholds to show the allocation/coefficients of metabolites</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.biomass_gene_essentiality">
<span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">biomass_gene_essentiality</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ko_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sel_para</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'zscore'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">th</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.biomass_gene_essentiality" title="Permalink to this definition">¶</a></dt>
<dd><section id="id1">
<h2>Description<a class="headerlink" href="#id1" title="Permalink to this heading">¶</a></h2>
<ol class="arabic simple">
<li><p>Calculate the changes of biomass objective values with different KO</p></li>
<li><p>Identify genes resulting in significantly changes with diff criteria</p></li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">param ko_df (pandas.DataFrame)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type ko_df (pandas.DataFrame)<span class="colon">:</span></dt>
<dd class="field-even"><p>data table–&gt;genes–&gt;columns and rxns–&gt;rows</p>
</dd>
<dt class="field-odd">param sel_para<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">param method (str)<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
<dt class="field-odd">type method (str)<span class="colon">:</span></dt>
<dd class="field-odd"><p>the criteria used to identify genes</p>
</dd>
<dt class="field-even">param - ‘zscore’<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
<dt class="field-odd">type - ‘zscore’<span class="colon">:</span></dt>
<dd class="field-odd"><p>default, standarize data and find genes with provided thresholds</p>
</dd>
<dt class="field-even">param - ‘percentage’<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
<dt class="field-odd">type - ‘percentage’<span class="colon">:</span></dt>
<dd class="field-odd"><p>gene is essential if KO biomass&lt;N% WT biomass (Yang et al., 2014)</p>
</dd>
<dt class="field-even">param th (float)<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
<dt class="field-odd">type th (float)<span class="colon">:</span></dt>
<dd class="field-odd"><p>thresholds for identifying genes, num of std/percentage for zscore/percentage</p>
</dd>
<dt class="field-even">returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>sigGenes (list)</strong></p>
</dd>
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><p>a list of significant genes</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.boxplot_fig">
<span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">boxplot_fig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mets_category</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value_ordering</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pv</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col3</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">portion</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plottype</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'boxplot'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Normalized</span> <span class="pre">coefficients'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_root_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/home/daweilin/StemCell/Project_mESC_JinZhang/regressor_results/'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.boxplot_fig" title="Permalink to this definition">¶</a></dt>
<dd><section id="id2">
<h2>Description<a class="headerlink" href="#id2" title="Permalink to this heading">¶</a></h2>
<ul class="simple">
<li><p>Identifies significant metabolites among up to three different samples</p></li>
<li><p>Calculates based on p-values and fold changes</p></li>
<li><p>Generates figures</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">param input_df (pandas.DataFrame)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type input_df (pandas.DataFrame)<span class="colon">:</span></dt>
<dd class="field-even"><p>a table with metabolites as rows and cells as columns</p>
</dd>
<dt class="field-odd">param mets_category (dictionary)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type mets_category (dictionary)<span class="colon">:</span></dt>
<dd class="field-even"><p>a dict of metabolites with corresponding types</p>
</dd>
<dt class="field-odd">param labels (list/pandas.DataFrame/numpy.array)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type labels (list/pandas.DataFrame/numpy.array)<span class="colon">:</span></dt>
<dd class="field-even"><p>desired labels for columns</p>
</dd>
<dt class="field-odd">param col_str (str)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type col_str (str)<span class="colon">:</span></dt>
<dd class="field-even"><p>name of the reference column used to sort values</p>
</dd>
<dt class="field-odd">param col1 (str)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type col1 (str)<span class="colon">:</span></dt>
<dd class="field-even"><p>name of the first sample (e.g. control)</p>
</dd>
<dt class="field-odd">param col2 (str)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type col2 (str)<span class="colon">:</span></dt>
<dd class="field-even"><p>name of the second sample (e.g. experiment)</p>
</dd>
<dt class="field-odd">param prefix (str)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type prefix (str)<span class="colon">:</span></dt>
<dd class="field-even"><p>name of the experiment for saving figures</p>
</dd>
<dt class="field-odd">param Optional Arguments<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">param ——————<span class="colon">:</span></dt>
<dd class="field-even"><p></p></dd>
<dt class="field-odd">param value_ordering (bool)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type value_ordering (bool)<span class="colon">:</span></dt>
<dd class="field-even"><p>sort values based on fold changes if set true</p>
</dd>
<dt class="field-odd">param pv (float)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type pv (float)<span class="colon">:</span></dt>
<dd class="field-even"><p>p-value thresholds</p>
</dd>
<dt class="field-odd">param fc (float)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type fc (float)<span class="colon">:</span></dt>
<dd class="field-even"><p>fold change thresholds</p>
</dd>
<dt class="field-odd">param col3 (str)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type col3 (str)<span class="colon">:</span></dt>
<dd class="field-even"><p>name of the third sample</p>
</dd>
<dt class="field-odd">param portion (float)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type portion (float)<span class="colon">:</span></dt>
<dd class="field-even"><p>the portion of cells selecting a metabolite</p>
</dd>
<dt class="field-odd">param norm (bool)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type norm (bool)<span class="colon">:</span></dt>
<dd class="field-even"><p>scale values to the range between [0, 1] if set True</p>
</dd>
<dt class="field-odd">param plottype (str)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type plottype (str)<span class="colon">:</span></dt>
<dd class="field-even"><p>boxplot or stripplot with mean values</p>
</dd>
<dt class="field-odd">param xlabels (str)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type xlabels (str)<span class="colon">:</span></dt>
<dd class="field-even"><p>name of the label on the x-axis</p>
</dd>
<dt class="field-odd">param save_root_path (str)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type save_root_path (str)<span class="colon">:</span></dt>
<dd class="field-even"><p>path to save the figure</p>
</dd>
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>coef_boxplot(dfs) (function)</strong></p>
</dd>
<dt class="field-even">rtype<span class="colon">:</span></dt>
<dd class="field-even"><p>generate a figure and dataframe of calculations</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.clustering_func">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">clustering_func</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pca_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mets_category</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.clustering_func" title="Permalink to this definition">¶</a></dt>
<dd><section id="parameter">
<h2>Parameter<a class="headerlink" href="#parameter" title="Permalink to this heading">¶</a></h2>
<p>pca_df (pandas dataframe): genes/reactions/objectives by cells/tissues IDs</p>
<dl class="field-list simple">
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>res_df (pandas dataframe)</strong></p>
</dd>
<dt class="field-even">rtype<span class="colon">:</span></dt>
<dd class="field-even"><p>pca_df with new columns that records entropy and clusters</p>
</dd>
</dl>
<dl class="py class">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.clustering_func.StandardScaler">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">StandardScaler</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_mean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_std</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.clustering_func.StandardScaler" title="Permalink to this definition">¶</a></dt>
<dd><p>Standardize features by removing the mean and scaling to unit variance.</p>
<p>The standard score of a sample <cite>x</cite> is calculated as:</p>
<blockquote>
<div><p>z = (x - u) / s</p>
</div></blockquote>
<p>where <cite>u</cite> is the mean of the training samples or zero if <cite>with_mean=False</cite>,
and <cite>s</cite> is the standard deviation of the training samples or one if
<cite>with_std=False</cite>.</p>
<p>Centering and scaling happen independently on each feature by computing
the relevant statistics on the samples in the training set. Mean and
standard deviation are then stored to be used on later data using
<a class="reference internal" href="#SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.transform" title="SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.transform"><code class="xref py py-meth docutils literal notranslate"><span class="pre">transform()</span></code></a>.</p>
<p>Standardization of a dataset is a common requirement for many
machine learning estimators: they might behave badly if the
individual features do not more or less look like standard normally
distributed data (e.g. Gaussian with 0 mean and unit variance).</p>
<p>For instance many elements used in the objective function of
a learning algorithm (such as the RBF kernel of Support Vector
Machines or the L1 and L2 regularizers of linear models) assume that
all features are centered around 0 and have variance in the same
order. If a feature has a variance that is orders of magnitude larger
that others, it might dominate the objective function and make the
estimator unable to learn from other features correctly as expected.</p>
<p>This scaler can also be applied to sparse CSR or CSC matrices by passing
<cite>with_mean=False</cite> to avoid breaking the sparsity structure of the data.</p>
<p>Read more in the <span class="xref std std-ref">User Guide</span>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>copy</strong> (<em>bool</em><em>, </em><em>default=True</em>) – If False, try to avoid a copy and do inplace scaling instead.
This is not guaranteed to always work inplace; e.g. if the data is
not a NumPy array or scipy.sparse CSR matrix, a copy may still be
returned.</p></li>
<li><p><strong>with_mean</strong> (<em>bool</em><em>, </em><em>default=True</em>) – If True, center the data before scaling.
This does not work (and will raise an exception) when attempted on
sparse matrices, because centering them entails building a dense
matrix which in common use cases is likely to be too large to fit in
memory.</p></li>
<li><p><strong>with_std</strong> (<em>bool</em><em>, </em><em>default=True</em>) – If True, scale the data to unit variance (or equivalently,
unit standard deviation).</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.scale_">
<span class="sig-name descname"><span class="pre">scale_</span></span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.scale_" title="Permalink to this definition">¶</a></dt>
<dd><p>Per feature relative scaling of the data to achieve zero mean and unit
variance. Generally this is calculated using <cite>np.sqrt(var_)</cite>. If a
variance is zero, we can’t achieve unit variance, and the data is left
as-is, giving a scaling factor of 1. <cite>scale_</cite> is equal to <cite>None</cite>
when <cite>with_std=False</cite>.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.17: </span><em>scale_</em></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>ndarray of shape (n_features,) or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.mean_">
<span class="sig-name descname"><span class="pre">mean_</span></span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.mean_" title="Permalink to this definition">¶</a></dt>
<dd><p>The mean value for each feature in the training set.
Equal to <code class="docutils literal notranslate"><span class="pre">None</span></code> when <code class="docutils literal notranslate"><span class="pre">with_mean=False</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>ndarray of shape (n_features,) or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.var_">
<span class="sig-name descname"><span class="pre">var_</span></span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.var_" title="Permalink to this definition">¶</a></dt>
<dd><p>The variance for each feature in the training set. Used to compute
<cite>scale_</cite>. Equal to <code class="docutils literal notranslate"><span class="pre">None</span></code> when <code class="docutils literal notranslate"><span class="pre">with_std=False</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>ndarray of shape (n_features,) or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.n_features_in_">
<span class="sig-name descname"><span class="pre">n_features_in_</span></span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.n_features_in_" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of features seen during <span class="xref std std-term">fit</span>.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.24.</span></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.feature_names_in_">
<span class="sig-name descname"><span class="pre">feature_names_in_</span></span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.feature_names_in_" title="Permalink to this definition">¶</a></dt>
<dd><p>Names of features seen during <span class="xref std std-term">fit</span>. Defined only when <cite>X</cite>
has feature names that are all strings.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 1.0.</span></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>ndarray of shape (<cite>n_features_in_</cite>,)</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.n_samples_seen_">
<span class="sig-name descname"><span class="pre">n_samples_seen_</span></span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.n_samples_seen_" title="Permalink to this definition">¶</a></dt>
<dd><p>The number of samples processed by the estimator for each feature.
If there are no missing samples, the <code class="docutils literal notranslate"><span class="pre">n_samples_seen</span></code> will be an
integer, otherwise it will be an array of dtype int. If
<cite>sample_weights</cite> are used it will be a float (if no missing data)
or an array of dtype float that sums the weights seen so far.
Will be reset on new calls to fit, but increments across
<code class="docutils literal notranslate"><span class="pre">partial_fit</span></code> calls.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int or ndarray of shape (n_features,)</p>
</dd>
</dl>
</dd></dl>

<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">scale</span></code></dt><dd><p>Equivalent function without the estimator API.</p>
</dd>
<dt><code class="xref py py-class docutils literal notranslate"><span class="pre">PCA</span></code></dt><dd><p>Further removes the linear correlation across features with ‘whiten=True’.</p>
</dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>NaNs are treated as missing values: disregarded in fit, and maintained in
transform.</p>
<p>We use a biased estimator for the standard deviation, equivalent to
<cite>numpy.std(x, ddof=0)</cite>. Note that the choice of <cite>ddof</cite> is unlikely to
affect model performance.</p>
<p>For a comparison of the different scalers, transformers, and normalizers,
see <span class="xref std std-ref">examples/preprocessing/plot_all_scaling.py</span>.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">sklearn.preprocessing</span> <span class="kn">import</span> <span class="n">StandardScaler</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="p">[[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scaler</span> <span class="o">=</span> <span class="n">StandardScaler</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">scaler</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">data</span><span class="p">))</span>
<span class="go">StandardScaler()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">scaler</span><span class="o">.</span><span class="n">mean_</span><span class="p">)</span>
<span class="go">[0.5 0.5]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">scaler</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">data</span><span class="p">))</span>
<span class="go">[[-1. -1.]</span>
<span class="go"> [-1. -1.]</span>
<span class="go"> [ 1.  1.]</span>
<span class="go"> [ 1.  1.]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">scaler</span><span class="o">.</span><span class="n">transform</span><span class="p">([[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">]]))</span>
<span class="go">[[3. 3.]]</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.fit">
<span class="sig-name descname"><span class="pre">fit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the mean and std to be used for later scaling.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>{array-like</em><em>, </em><em>sparse matrix}</em><em> of </em><em>shape</em><em> (</em><em>n_samples</em><em>, </em><em>n_features</em><em>)</em>) – The data used to compute the mean and standard deviation
used for later scaling along the features axis.</p></li>
<li><p><strong>y</strong> (<em>None</em>) – Ignored.</p></li>
<li><p><strong>sample_weight</strong> (<em>array-like</em><em> of </em><em>shape</em><em> (</em><em>n_samples</em><em>,</em><em>)</em><em>, </em><em>default=None</em>) – <p>Individual weights for each sample.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.24: </span>parameter <em>sample_weight</em> support to StandardScaler.</p>
</div>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>self</strong> – Fitted scaler.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>object</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.inverse_transform">
<span class="sig-name descname"><span class="pre">inverse_transform</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.inverse_transform" title="Permalink to this definition">¶</a></dt>
<dd><p>Scale back the data to the original representation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>{array-like</em><em>, </em><em>sparse matrix}</em><em> of </em><em>shape</em><em> (</em><em>n_samples</em><em>, </em><em>n_features</em><em>)</em>) – The data used to scale along the features axis.</p></li>
<li><p><strong>copy</strong> (<em>bool</em><em>, </em><em>default=None</em>) – Copy the input X or not.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>X_tr</strong> – Transformed array.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>{ndarray, sparse matrix} of shape (n_samples, n_features)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.partial_fit">
<span class="sig-name descname"><span class="pre">partial_fit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.partial_fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Online computation of mean and std on X for later scaling.</p>
<p>All of X is processed as a single batch. This is intended for cases
when <a class="reference internal" href="#SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.fit" title="SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.fit"><code class="xref py py-meth docutils literal notranslate"><span class="pre">fit()</span></code></a> is not feasible due to very large number of
<cite>n_samples</cite> or because X is read from a continuous stream.</p>
<p>The algorithm for incremental mean and std is given in Equation 1.5a,b
in Chan, Tony F., Gene H. Golub, and Randall J. LeVeque. “Algorithms
for computing the sample variance: Analysis and recommendations.”
The American Statistician 37.3 (1983): 242-247:</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>{array-like</em><em>, </em><em>sparse matrix}</em><em> of </em><em>shape</em><em> (</em><em>n_samples</em><em>, </em><em>n_features</em><em>)</em>) – The data used to compute the mean and standard deviation
used for later scaling along the features axis.</p></li>
<li><p><strong>y</strong> (<em>None</em>) – Ignored.</p></li>
<li><p><strong>sample_weight</strong> (<em>array-like</em><em> of </em><em>shape</em><em> (</em><em>n_samples</em><em>,</em><em>)</em><em>, </em><em>default=None</em>) – <p>Individual weights for each sample.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.24: </span>parameter <em>sample_weight</em> support to StandardScaler.</p>
</div>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>self</strong> – Fitted scaler.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>object</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.transform">
<span class="sig-name descname"><span class="pre">transform</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.clustering_func.StandardScaler.transform" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform standardization by centering and scaling.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>{array-like</em><em>, </em><em>sparse matrix</em><em> of </em><em>shape</em><em> (</em><em>n_samples</em><em>, </em><em>n_features</em><em>)</em>) – The data used to scale along the features axis.</p></li>
<li><p><strong>copy</strong> (<em>bool</em><em>, </em><em>default=None</em>) – Copy the input X or not.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>X_tr</strong> – Transformed array.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>{ndarray, sparse matrix} of shape (n_samples, n_features)</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.clustering_func.reclustering">
<span class="sig-name descname"><span class="pre">reclustering</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cluster_res</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'UMAP'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title_suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.clustering_func.reclustering" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>(</strong><strong>pandas.DataFrame</strong><strong>)</strong> (<em>cluster_res</em>) – </p></li>
<li><p><strong>(</strong><strong>list</strong><strong>)</strong> (<em>cols</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>recluster_res (pandas.DataFrame)</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>data reclustering coming with the true/predicted labels.</p>
</dd>
</dl>
</dd></dl>

</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.coef_distance_to_biomassObj">
<span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">coef_distance_to_biomassObj</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biomass_coef_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'euclidean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zscore</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_input_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'model'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_root_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/home/daweilin/StemCell/Project_mESC_JinZhang/regressor_results_new/'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">histplot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">boxplot_cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">boxplot_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.coef_distance_to_biomassObj" title="Permalink to this definition">¶</a></dt>
<dd><p>Visualize Eucleadian distances from inferred objectives to biomass objectives</p>
<p>Objective coefficients of up to three different samples will be compared to the coefficients
in the biomass objectives metabolites with Eucleadian distances and statistically tested with T-test.
Note that the coefficients of h2o is ignored by this function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>{pandas.DataFrame}</em><em> of </em><em>shape</em><em> (</em><em>n_metabolites</em><em>, </em><em>n_samples</em><em>)</em>) – a table with metabolites as rows and cells as columns</p></li>
<li><p><strong>labels</strong> (<em>{list/pandas.DataFrame/numpy.array}</em><em>,</em>) – desired labels for columns</p></li>
<li><p><strong>biomass_coef_df</strong> (<em>{pandas.DataFrame}</em><em> of </em><em>shape</em><em> (</em><em>n_metabolites</em><em>, </em><em>1</em><em>)</em>) – table that saves coefficients of a biomass objective that could be from Recon1, Recon2.2, or Recon3D</p></li>
<li><p><strong>prefix</strong> (<em>{str}</em>) – name of the experiment for saving figures</p></li>
<li><p><strong>func</strong> (<em>{string}</em><em>, </em><em>default='euclidean'</em>) – name of the method used for the calculation of distances between coefficients</p></li>
<li><p><strong>norm</strong> (<em>{bool}</em><em>, </em><em>default=False</em>) – normalize the coefficient of each sample by the summation of coefficients of metabolites</p></li>
<li><p><strong>save_root_path</strong> (<em>{str}</em><em>, </em><em>default='/home/daweilin/StemCell/Project_mESC_JinZhang/regressor_results/'</em>) – path to save the figure</p></li>
<li><p><strong>histplot</strong> (<em>{bool}</em><em>, </em><em>default=True</em>) – show histogram next to the scatterplot</p></li>
<li><p><strong>boxplot_cols</strong> (<em>{array-like}</em><em>, </em><em>default=</em><em>[</em><em>]</em>) – [WARNING] this input has been deprecated. will be removed in the future</p></li>
<li><p><strong>boxplot_order</strong> (<em>{array-like}</em><em>, </em><em>default=</em><em>[</em><em>]</em>) – names of samples corresponding to labels that order the scatterplot from top to bottom</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.constrained_models">
<span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">constrained_models</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">root_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">CFR_paraScan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">DFA_paraScan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">randomScan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'_fluxes.csv.gz'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_path_pattern</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biomass_reaction_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'biomass_objective'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">CFR_k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[10,</span> <span class="pre">1,</span> <span class="pre">0.1,</span> <span class="pre">0.01,</span> <span class="pre">0.001]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">CFR_r</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[10,</span> <span class="pre">1,</span> <span class="pre">0.1,</span> <span class="pre">0.01,</span> <span class="pre">0.001]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">DFA_k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[10,</span> <span class="pre">1,</span> <span class="pre">0.1,</span> <span class="pre">0.01,</span> <span class="pre">0.001]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.constrained_models" title="Permalink to this definition">¶</a></dt>
<dd><section id="id3">
<h2>Description<a class="headerlink" href="#id3" title="Permalink to this heading">¶</a></h2>
<p>Organize constrained modeling of fluxes without objectives</p>
<dl class="field-list simple">
<dt class="field-odd">param root_path (string)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type root_path (string)<span class="colon">:</span></dt>
<dd class="field-even"><p>path to access all the files</p>
</dd>
<dt class="field-odd">param CFR_paraScan (list of float)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type CFR_paraScan (list of float)<span class="colon">:</span></dt>
<dd class="field-even"><p>CFR parameters of interest used to output fluxes</p>
</dd>
<dt class="field-odd">param DFA_paraScan (list of float)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type DFA_paraScan (list of float)<span class="colon">:</span></dt>
<dd class="field-even"><p>DFA parameters of interest used to output fluxes</p>
</dd>
<dt class="field-odd">param randomScan (bool)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type randomScan (bool)<span class="colon">:</span></dt>
<dd class="field-even"><p>read all models without considering parameters</p>
</dd>
<dt class="field-odd">param norm (bool)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type norm (bool)<span class="colon">:</span></dt>
<dd class="field-even"><p>if normalize the data by the summation of fluxes or not</p>
</dd>
<dt class="field-odd">param file_suffix (string)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type file_suffix (string)<span class="colon">:</span></dt>
<dd class="field-even"><p>the suffix of filenames to read</p>
</dd>
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>res_df (pandas.DataFrame)</strong></p>
</dd>
<dt class="field-even">rtype<span class="colon">:</span></dt>
<dd class="field-even"><p>reaction fluxes with parameters of interest</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.extracellular_flux_matching">
<span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">extracellular_flux_matching</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metabolome_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sheet_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">uncon_res</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mets_category</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">MetCol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Extracellular</span> <span class="pre">Metabolites'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.extracellular_flux_matching" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>(</strong><strong>str</strong><strong>)</strong> (<em>sheet_name</em>) – </p></li>
<li><p><strong>(</strong><strong>str</strong><strong>)</strong> – </p></li>
<li><p><strong>(</strong><strong>pandas.DataFrame</strong><strong>)</strong> (<em>uncon_res</em>) – </p></li>
<li><p><strong>(</strong><strong>dictionary</strong><strong>)</strong> (<em>mets_category</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>match_map_df (pandas.DataFrame)</em></p></li>
<li><p><em>ffm_cp (pandas.DataFrame)</em></p></li>
<li><p><em>ffm_reg (pandas.DataFrame)</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.fva_constrained_models">
<span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">fva_constrained_models</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">root_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.fva_constrained_models" title="Permalink to this definition">¶</a></dt>
<dd><p>Data loading: constrained modeling of HumanAtlas without objectives</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.getRxnCoefficients">
<span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">getRxnCoefficients</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/home/daweilin/StemCell/cancer_model.mat'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Recon1'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.getRxnCoefficients" title="Permalink to this definition">¶</a></dt>
<dd><p>Breakdown coefficients of biomass objectives from a given metabolic model</p>
<p>A biomass objective is a linear combination of multiple metabolites as reactants with non-negative coefficients.
Although there would be products produced by a biomass objective reaction, the function will ignore them.
The biomass objective could be from different metabolic models such as Recon1, Recon2.2, and Recon3D.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model_path</strong> (<em>{string}</em><em>, </em><em>default='/home/daweilin/StemCell/cancer_model.mat'</em>) – path to access a metabolic model</p></li>
<li><p><strong>model_name</strong> (<em>{string}</em><em>, </em><em>default='Recon1'</em>) – name of the metabolic model</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>biomass_coef_df</strong> – table that saves metabolites in rows and corresponding coefficients in the first column</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>{pandas.DataFrame} of shape (n_metabolites, 1)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.get_MSigDB">
<span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">get_MSigDB</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dbName</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'2023.1.Hs'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">categoryName</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'h.all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">list_db</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">list_category</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_to_csv</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flip</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.get_MSigDB" title="Permalink to this definition">¶</a></dt>
<dd><p>Access to significant genes collected in MSigDB database</p>
<p>This function relies on the GESApy package to access MSigDB and retrieve
certain categories of significant genes from an assigned database.
The releases of the data are from
<a class="reference external" href="https://data.broadinstitute.org/gsea-msigdb/msigdb/release/">https://data.broadinstitute.org/gsea-msigdb/msigdb/release/</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dbName</strong> (<em>{string}</em><em>, </em><em>default=&quot;2023.1.Hs&quot;</em>) – Name of MSigDB database. Please check the url provided above.
(Hs=Human, Mm=mouse)</p></li>
<li><p><strong>categoryName</strong> (<em>{string}</em><em>, </em><em>default='h.all'</em>) – Name of categories of significant genes.</p></li>
<li><p><strong>list_db</strong> (<em>{bool}</em><em>, </em><em>default=False</em>) – helper function to check out all the databases</p></li>
<li><p><strong>list_category</strong> (<em>{bool}</em><em>, </em><em>default=False</em>) – helper function to check out all the categories in a given database</p></li>
<li><p><strong>save_to_csv</strong> (<em>{string}</em><em>, </em><em>default=''</em>) – if the parameter is not empty, the significant genes will be saved in
.csv files with filenames that are name of gene sets.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>gmt</strong> – Name of gene sets and corresponding significant genes</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>{dictionary}, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.hypergeom_test">
<span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">hypergeom_test</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">M</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">N</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.hypergeom_test" title="Permalink to this definition">¶</a></dt>
<dd><p>arr1: array with only 1, 0, -1
arr2: array with only 1, 0, -1</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.load_multiObj_models">
<span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">load_multiObj_models</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">root_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'KSOM'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_variables</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">CFR_paraScan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">DFA_paraScan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">randomScan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">topology_use</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'_fluxes.csv.gz'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ind_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biomass_reaction_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'biomass_objective'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">CFR_k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[10,</span> <span class="pre">1,</span> <span class="pre">0.1,</span> <span class="pre">0.01,</span> <span class="pre">0.001]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">CFR_r</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[10,</span> <span class="pre">1,</span> <span class="pre">0.1,</span> <span class="pre">0.01,</span> <span class="pre">0.001]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">DFA_k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[10,</span> <span class="pre">1,</span> <span class="pre">0.1,</span> <span class="pre">0.01,</span> <span class="pre">0.001]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.load_multiObj_models" title="Permalink to this definition">¶</a></dt>
<dd><section id="id4">
<h2>Description<a class="headerlink" href="#id4" title="Permalink to this heading">¶</a></h2>
<p>Organize any types of models</p>
<dl class="field-list simple">
<dt class="field-odd">param root_path (string)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type root_path (string)<span class="colon">:</span></dt>
<dd class="field-even"><p>path to access all the files</p>
</dd>
<dt class="field-odd">param medium (string)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type medium (string)<span class="colon">:</span></dt>
<dd class="field-even"><p>medium used for input metabolites e.g. DMEMF12 or KSOM</p>
</dd>
<dt class="field-odd">param return_variables (bool)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type return_variables (bool)<span class="colon">:</span></dt>
<dd class="field-even"><p>True for return fluxes or False for objective values</p>
</dd>
<dt class="field-odd">param norm (bool)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type norm (bool)<span class="colon">:</span></dt>
<dd class="field-even"><p>if normalize the data by the summation of fluxes or not</p>
</dd>
<dt class="field-odd">param CFR_paraScan (bool)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type CFR_paraScan (bool)<span class="colon">:</span></dt>
<dd class="field-even"><p>if filtering models with CFR parameters of interest</p>
</dd>
<dt class="field-odd">param DFA_paraScan (bool)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type DFA_paraScan (bool)<span class="colon">:</span></dt>
<dd class="field-even"><p>if filtering models with DFA parameters of interest</p>
</dd>
<dt class="field-odd">param randomScan (bool)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type randomScan (bool)<span class="colon">:</span></dt>
<dd class="field-even"><p>read all models without considering parameters</p>
</dd>
<dt class="field-odd">param topology_use (bool)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type topology_use (bool)<span class="colon">:</span></dt>
<dd class="field-even"><p>remove fluxes of candidate metabolites but biomass objective</p>
</dd>
<dt class="field-odd">param CFR_k (list of floats)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type CFR_k (list of floats)<span class="colon">:</span></dt>
<dd class="field-even"><p>CFR kappa of interest used to output fluxes</p>
</dd>
<dt class="field-odd">param CFR_r (list of floats)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type CFR_r (list of floats)<span class="colon">:</span></dt>
<dd class="field-even"><p>CFR rho of interest used to output fluxes</p>
</dd>
<dt class="field-odd">param CFR_r (list of floats)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type CFR_r (list of floats)<span class="colon">:</span></dt>
<dd class="field-even"><p>DFA parameters of interest used to output fluxes</p>
</dd>
<dt class="field-odd">param file_suffix (string)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type file_suffix (string)<span class="colon">:</span></dt>
<dd class="field-even"><p>the suffix of filenames to read</p>
</dd>
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>res_df (pandas.DataFrame)</strong></p>
</dd>
<dt class="field-even">rtype<span class="colon">:</span></dt>
<dd class="field-even"><p>reaction fluxes with parameters of interest</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.logVariance_allocation">
<span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">logVariance_allocation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sampleName</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replace_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dataType</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Objective'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/home/daweilin/StemCell/cancer_model.mat'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.logVariance_allocation" title="Permalink to this definition">¶</a></dt>
<dd><section id="id5">
<h2>Parameter<a class="headerlink" href="#id5" title="Permalink to this heading">¶</a></h2>
<p>flux_df (pandas.DataFrame): table of reaction fluxes
obj_df (pandas.DataFrame): table of objective coefficients
sampleName (str): name of the sample which will be used for the prefix of filenames (e.g. scEmbryo)
replace_dict (dictionary): e.g. {‘1Cto2cell’:’sc1C2C’, ‘2Cto32cell’:’sc2CBC’}</p>
<dl class="field-list simple">
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.lollipop_fig">
<span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">lollipop_fig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mets_category</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value_ordering</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_root_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/home/daweilin/StemCell/Project_mESC_JinZhang/regressor_results/'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.lollipop_fig" title="Permalink to this definition">¶</a></dt>
<dd><p>Visualize metabolic objectives with lollipop plots</p>
<ol class="arabic simple">
<li><p>Visualize proportion of metabolites selected by up to three different samples</p></li>
<li><p>Generates figures</p></li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_df</strong> (<em>{pandas.DataFrame}</em>) – a table with metabolites as rows and cells as columns</p></li>
<li><p><strong>mets_category</strong> (<em>{dictionary}</em>) – a dict of metabolites with corresponding types</p></li>
<li><p><strong>labels</strong> (<em>{list/pandas.DataFrame/numpy.array}</em><em>,</em>) – desired labels for columns</p></li>
<li><p><strong>cols</strong> (<em>{list/numpy.array}</em>) – names of the samples (e.g. control, exp1, exp2)</p></li>
<li><p><strong>prefix</strong> (<em>{str}</em>) – name of the experiment for saving figures</p></li>
<li><p><strong>value_ordering</strong> (<em>{bool}</em><em>, </em><em>default=True</em>) – sort values based on fold changes if set true</p></li>
<li><p><strong>cutoff</strong> (<em>{float}</em><em>, </em><em>default=0.1</em>) – the portion of cells selecting a metabolite</p></li>
<li><p><strong>save_root_path</strong> (<em>{str}</em><em>, </em><em>default='/home/daweilin/StemCell/Project_mESC_JinZhang/regressor_results/'</em>) – path to save the figure</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>plot_df</strong> – generate a figure and dataframe of calculations</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>{pandas.DataFrame}</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.metObj_gene_essentiality">
<span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">metObj_gene_essentiality</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">coef_paths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sel_para</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'zscore'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">th</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.metObj_gene_essentiality" title="Permalink to this definition">¶</a></dt>
<dd><p># Process objective coefficients
coef_paths = {</p>
<blockquote>
<div><p>‘ESC’:f’/nfs/turbo/umms-csriram/daweilin/regression_models/GeneKO/flux_sl_ESC_{sel_para}_norm_nonorm.csv’,
}</p>
</div></blockquote>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.objective_flux_matching">
<span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">objective_flux_matching</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metabolome_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sheet_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">MetCol</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mets_category</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">manual_map</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zscore_mets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.objective_flux_matching" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>(</strong><strong>str</strong><strong>)</strong> (<em>MetCol</em>) – </p></li>
<li><p><strong>(</strong><strong>str</strong><strong>)</strong> – </p></li>
<li><p><strong>(</strong><strong>str</strong><strong>)</strong> – </p></li>
<li><p><strong>(</strong><strong>dictionary</strong><strong>)</strong> (<em>mets_category</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>match_map_df (pandas.DataFrame)</em></p></li>
<li><p><em>ffm_cp (pandas.DataFrame)</em></p></li>
<li><p><em>ffm_reg (pandas.DataFrame)</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.overlap_allocation_plot">
<span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">overlap_allocation_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">coef_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.overlap_allocation_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>The allocation of each metabolite in objectives that are shown with overlapped barplots.</p>
<p>The allocation of coefficients will be represented by
c/Sigma(c) if norm is set true</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>coef_df</strong> (<em>{pandas.DataFrame}</em><em> of </em><em>shape</em><em> (</em><em>n_metabolites</em><em>, </em><em>n_samples</em><em>)</em><em>,</em>) – the coefficients of metabolic objectives for each samples</p></li>
<li><p><strong>labels</strong> (<em>{array-like}</em><em> of </em><em>shape</em><em> (</em><em>n_samples</em><em>,</em><em>)</em><em>,</em>) – the labels of samples/columns</p></li>
<li><p><strong>cellType</strong> (<em>{string}</em><em>,</em>) – the label used as a reference</p></li>
<li><p><strong>prefix</strong> (<em>{string}</em><em>, </em><em>default=''</em>) – experiment names for saving files</p></li>
<li><p><strong>norm</strong> (<em>{bool}</em><em>, </em><em>default=True</em>) – normalize coefficients by the sum of all the coefficients</p></li>
<li><p><strong>cutoff</strong> (<em>{float}</em><em>, </em><em>default=0.0</em>) – thresholds to show the allocation/coefficients of metabolites</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.read_mat_model">
<span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">read_mat_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/home/daweilin/StemCell/cancer_model.mat'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_rxnGeneMat</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_transport</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.read_mat_model" title="Permalink to this definition">¶</a></dt>
<dd><section id="id6">
<h2>Description<a class="headerlink" href="#id6" title="Permalink to this heading">¶</a></h2>
<ol class="arabic simple">
<li><p>read genome-scale metabolic network models (GEMs) from .mat file</p></li>
<li><p>reorganize the model and save into a python dictionary</p></li>
<li><p>(optional) gather subsystems, reacitons, and genes with gpr rules</p></li>
<li><p>(optional) remove transport and exchange reactions for internal reactions analysis</p></li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">param path (string)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type path (string)<span class="colon">:</span></dt>
<dd class="field-even"><p>the path to access the network model</p>
</dd>
<dt class="field-odd">param return_rxnGeneMat (bool)<span class="colon">:</span></dt>
<dd class="field-odd"><p>where columns are genes, additional columns for reactions and subsystems</p>
</dd>
<dt class="field-even">type return_rxnGeneMat (bool)<span class="colon">:</span></dt>
<dd class="field-even"><p>if output a binary dataframe showing association of GPR rules</p>
</dd>
<dt class="field-odd">param remove_transport (bool)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type remove_transport (bool)<span class="colon">:</span></dt>
<dd class="field-even"><p>if remove reactions related to exchange/transport subsystems</p>
</dd>
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model_dict (dictionary)</strong> (<em>fields of models and corresponding values, e.g. ‘genes’:[‘A’, ‘B’…]</em>)</p></li>
<li><p><strong>rxnGeneMat (optional, pandas.DataFrame)</strong> (<em>a dataframe with binary values showing association</em>) – between genes and reactions where column names are genes,
additional two columns for reactions and subsystems</p></li>
</ul>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SCOOTI.regressionAnalyzer.unconstrained_models">
<span class="sig-prename descclassname"><span class="pre">SCOOTI.regressionAnalyzer.</span></span><span class="sig-name descname"><span class="pre">unconstrained_models</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">root_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_variables</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'DMEMF12'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biomass_reaction_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'biomass_objective'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.regressionAnalyzer.unconstrained_models" title="Permalink to this definition">¶</a></dt>
<dd><section id="id7">
<h2>Description<a class="headerlink" href="#id7" title="Permalink to this heading">¶</a></h2>
<p>Unconstrained modeling of fluxes with different objectives</p>
<dl class="field-list simple">
<dt class="field-odd">param root_path (string)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type root_path (string)<span class="colon">:</span></dt>
<dd class="field-even"><p>path to access all the files</p>
</dd>
<dt class="field-odd">param norm (bool)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type norm (bool)<span class="colon">:</span></dt>
<dd class="field-even"><p>if normalize the data or not</p>
</dd>
<dt class="field-odd">param return_variables (bool)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type return_variables (bool)<span class="colon">:</span></dt>
<dd class="field-even"><p>True for return fluxes or False for objective values</p>
</dd>
<dt class="field-odd">param medium (string)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type medium (string)<span class="colon">:</span></dt>
<dd class="field-even"><p>medium used for input metabolites e.g. DMEMF12 or KSOM</p>
</dd>
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>res_df (optional, pandas.DataFrame)</strong> (<em>reaction fluxes excluding objective values</em>)</p></li>
<li><p><strong>dm_df (optional, pandas.DataFrame)</strong> (<em>fluxes of demand reactions of candidate metabolites</em>)</p></li>
</ul>
</dd>
</dl>
</section>
</dd></dl>

<span class="target" id="module-SCOOTI.regressorTraining"></span><section id="regressortraining-py">
<h2>regressorTraining.py<a class="headerlink" href="#regressortraining-py" title="Permalink to this heading">¶</a></h2>
<p>A module enables regression models to learn from metabolic fluxes</p>
</section>
<div class="toctree-wrapper compound">
</div>
</section>
<section id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this heading">¶</a></h1>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="#">SCOOTI</a></h1>








<h3>Navigation</h3>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="#">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2023, Chandrasekaran Lab, University of Michigan.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.3.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.13</a>
      
      |
      <a href="_sources/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>