
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Welcome to the documentation of metabolic objective analysis &#8212; SCOOTI 0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Quick Start" href="quickstart.html" />
    <link rel="prev" title="Welcome to the documentation of metabolic objective analysis" href="regressionAnalyzer.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="welcome-to-the-documentation-of-metabolic-objective-analysis">
<h1>Welcome to the documentation of metabolic objective analysis<a class="headerlink" href="#welcome-to-the-documentation-of-metabolic-objective-analysis" title="Permalink to this heading">¶</a></h1>
<section id="module-SCOOTI.metObjAnalyzer">
<span id="core-analysis-module"></span><h2>Core Analysis Module<a class="headerlink" href="#module-SCOOTI.metObjAnalyzer" title="Permalink to this heading">¶</a></h2>
<section id="metobjanalyzer-py">
<h3>metObjAnalyzer.py<a class="headerlink" href="#metobjanalyzer-py" title="Permalink to this heading">¶</a></h3>
<p>Object-oriented classes that serve as a user interface to run analysis/simulations</p>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">SCOOTI.metObjAnalyzer.</span></span><span class="sig-name descname"><span class="pre">metObjAnalyzer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">flux_paths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coef_paths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_root_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">GEM_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">uncon_model_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_map</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samplingFlux_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sel_para</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'KSOM'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">GEM_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer" title="Permalink to this definition">¶</a></dt>
<dd><p>Analysis of optimized fluxes and metabolic objectives.</p>
<p>The object consists of four parts of analysis.
1) Optimized flux:</p>
<blockquote>
<div><p>suggests the best parameters and visualization.</p>
</div></blockquote>
<ol class="arabic simple" start="2">
<li><dl class="simple">
<dt>Metabolic objective:</dt><dd><p>compare to different samples and allocations.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Tradeoffs and traits:</dt><dd><p>identify anticorrelated metabolites and metabolic traits.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Flux reconstruction:</dt><dd><p>reconstruct flux prediction with inferred coefficients.</p>
</dd>
</dl>
</li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>flux_paths</strong> (<em>{dictionary}</em><em>,</em>) – Paths to access the directories with flux prediction data.
Keys should be the labels (sample names) associated with the directories.</p></li>
<li><p><strong>coef_paths</strong> (<em>{dictionary}</em><em>,</em>) – Paths to access the files (.csv) with inferred metabolic objectives.
Keys should be the labels (sample names) associated with the files.</p></li>
<li><p><strong>save_root_path</strong> (<em>{str}</em><em>,</em>) – Path to save the outputs (including files and figures).
Absolute path is recommended.</p></li>
<li><p><strong>GEM_path</strong> (<em>{str}</em><em>,</em>) – Paths to access the genome-scale metabolic models (.mat files).
Only Shen et al., Recon2.2, and Recon3D are applicable for this version.</p></li>
<li><p><strong>uncon_model_path</strong> (<em>{str}</em><em>,</em>) – Paths to access the directories with flux predictions of single-objective models.</p></li>
<li><p><strong>col_map</strong> (<em>{dictionary}</em><em>,</em>) – Serialize unique column names (samples) and corresponding preferred names.</p></li>
<li><p><strong>label_func</strong> (<em>{function}</em><em>,</em>) – A function to convert column names into unique sample names.</p></li>
<li><p><strong>samplingFlux_path</strong> (<em>{str}</em><em>,</em>) – Paths to access the directories with flux predictions
models with randomly sampled objective coefficients.</p></li>
<li><p><strong>sel_para</strong> (<em>{str}</em><em>,</em>) – A parameter combination selected for CFR (e.g. k0.1_r0.01)</p></li>
<li><p><strong>prefix</strong> (<em>{str}</em><em>,</em>) – Project name (e.g. scEmbryo)</p></li>
<li><p><strong>medium</strong> (<em>{string}</em><em>, </em><em>options=</em><em>[</em><em>'KSOM'</em><em>, </em><em>'DMEMF12'</em><em>]</em>) – medium used to model the fluxes with FBA</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.flux_df">
<span class="sig-name descname"><span class="pre">flux_df</span></span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.flux_df" title="Permalink to this definition">¶</a></dt>
<dd><p>Flux data predicted with constrained models without objectives</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>{pandas.DataFrame},</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.coef_df">
<span class="sig-name descname"><span class="pre">coef_df</span></span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.coef_df" title="Permalink to this definition">¶</a></dt>
<dd><p>Coefficients of metabolic objectives</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>{pandas.DataFrame},</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.labels">
<span class="sig-name descname"><span class="pre">labels</span></span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.labels" title="Permalink to this definition">¶</a></dt>
<dd><p>label for each column of flux or coefficient data</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>{pandas.DataFrame},</p>
</dd>
</dl>
</dd></dl>

<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">SCOOTI.metObjAnalyzer</span> <span class="kn">import</span> <span class="n">metObjAnalyzer</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># set up a function to replace column names</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">def</span> <span class="nf">label_func</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">({</span><span class="s1">&#39;1C&#39;</span><span class="p">:</span><span class="s1">&#39;sc1C2C&#39;</span><span class="p">,</span> <span class="s1">&#39;2C&#39;</span><span class="p">:</span><span class="s1">&#39;sc2CBC&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># initiate the object</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">moa</span> <span class="o">=</span> <span class="n">metObjAnalyzer</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">flux_paths</span><span class="o">=</span><span class="p">{</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="s1">&#39;sc1C2C&#39;</span><span class="p">:</span><span class="s1">&#39;./fluxPrediction/sc1C2C/paraScan/&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="s1">&#39;sc2CBC&#39;</span><span class="p">:</span><span class="s1">&#39;./fluxPrediction/sc2CBC/paraScan/&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>                <span class="p">},</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">coef_paths</span><span class="o">=</span><span class="p">{</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="s1">&#39;sc1C2C&#39;</span><span class="p">:</span><span class="sa">f</span><span class="s1">&#39;./regression_models/scEmbryo_paraScan/flux_sl_sc1C2C_input_norm_outcome_nonorm_k0.1_r0.01.csv&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="s1">&#39;sc2CBC&#39;</span><span class="p">:</span><span class="sa">f</span><span class="s1">&#39;./regression_models/scEmbryo_paraScan/flux_sl_sc2CBC_input_norm_outcome_nonorm_k0.1_r0.01.csv&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">},</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">save_root_path</span><span class="o">=</span><span class="s1">&#39;/home/daweilin/StemCell/Project_mESC_JinZhang/regressor_results_new/&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">GEM_path</span><span class="o">=</span><span class="s1">&#39;./models/Recon1.mat&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">uncon_model_path</span><span class="o">=</span><span class="s1">&#39;./fluxPrediction/unconstrained_models/pfba/KSOM/&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">col_map</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;sc1C2C&#39;</span><span class="p">:</span><span class="s1">&#39;sc1C2C&#39;</span><span class="p">,</span> <span class="s1">&#39;sc2CBC&#39;</span><span class="p">:</span><span class="s1">&#39;sc2CBC&#39;</span><span class="p">},</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">label_func</span><span class="o">=</span><span class="n">label_func</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">samplingFlux_path</span><span class="o">=</span><span class="s1">&#39;./fluxPrediction/RandomObjCoef/embryo_ksom_50percent/&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">sel_para</span><span class="o">=</span><span class="s1">&#39;k0.1_r0.01&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;scEmbryo&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">medium</span><span class="o">=</span><span class="s1">&#39;KSOM&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Analyze fluxes and look for the best parameter combinations</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">moa</span><span class="o">.</span><span class="n">fluxAnalysis</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">kappa_arr</span><span class="o">=</span><span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">rho_arr</span><span class="o">=</span><span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># read coefficients</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">moa</span><span class="o">.</span><span class="n">get_coef</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># analyze coefficients</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coefAnalysis</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">clustering</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">entropy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">distance</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">umap_para</span><span class="o">=</span><span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">50</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="p">)</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.coefAnalysis">
<span class="sig-name descname"><span class="pre">coefAnalysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">norm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unknown_clustering</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clustering</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entropy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compare</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_ub</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umap_para</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[5,</span> <span class="pre">50]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recluster_min_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'average'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">special_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.coefAnalysis" title="Permalink to this definition">¶</a></dt>
<dd><p>Analysis of inferred metabolic objectives.</p>
<p>The coefficients of metabolic objectives were analyzed by
1) comparing to different samples,
2) dimension reduction methods for metabolic objectives,
3) objective entropy (degree of order),
4) distances from the biomass objective to the inferred objectives</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>norm</strong> (<em>{bool}</em><em>, </em><em>default=True</em>) – normalize the coefficients by the sum of all coefficients of a sample.</p></li>
<li><p><strong>clustering</strong> (<em>{bool}</em><em>, </em><em>default=True</em>) – enable dimension reduction methods or clustering methods for metabolic objectives.
UMAP is the default method in the class, but it is allowed to be swapped by
PCA, tSNE, and PHATE.</p></li>
<li><p><strong>entropy</strong> (<em>{bool}</em><em>, </em><em>default=True</em>) – enable the calculations of entropy based on the coefficients of metabolic objectives
which follows the equation derived from <cite>p(log(p))</cite>. <cite>p</cite> is converted from</p></li>
<li><p><strong>distance</strong> (<em>{bool}</em><em>, </em><em>default=True</em>) – enable dimension reduction methods or clustering methods for metabolic objectives</p></li>
<li><p><strong>umap_para</strong> (<em>{array-like}</em><em>, </em><em>default=</em><em>[</em><em>5</em><em>, </em><em>10</em><em>]</em>) – the first number of the array stands for the number of neighbors and
the second number of the array is the total dimension of UMAP</p></li>
<li><p><strong>recluster_min_size</strong> (<em>{int}</em><em>, </em><em>default=10</em>) – the parameter for HDBSCAN to cluster the UMAP results</p></li>
<li><p><strong>method</strong> (<em>{str}</em><em>, </em><em>default='average'</em>) – method used to cluster samples with scipy.cluster.hierarchy.linkage.
Options include ‘average’, ‘single’, ‘complete’, ‘weighted’, ‘centroid’.
For more details, one can visit the official document.
<a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.linkage.html">https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.linkage.html</a></p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>self</strong> – Returns self.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>object</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The figures of the analysis will be saved in the path given by <cite>self.save_root_path</cite>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.fluxAnalysis">
<span class="sig-name descname"><span class="pre">fluxAnalysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kappa</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rho</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umap_para</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[5,</span> <span class="pre">10]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.fluxAnalysis" title="Permalink to this definition">¶</a></dt>
<dd><p>Cluster fluxes of UMAP with HDBSCAN and output clustering perfomance.</p>
<p>Cell-specific constrained models without objective functions
1) The model predicts the fluxes of ideal optimality
2) We aim to model fluxes that can separate cell types</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>kappa_arr</strong> (<em>{array-like}</em><em> of </em><em>shape</em><em> (</em><em>n_parameters</em><em>,</em><em>)</em><em>, </em><em>default=</em><em>[</em><em>10</em><em>, </em><em>1</em><em>, </em><em>0.1</em><em>, </em><em>0.01</em><em>, </em><em>0.001</em><em>]</em>) – parameter of interest.</p></li>
<li><p><strong>rho_arr</strong> (<em>{array-like}</em><em> of </em><em>shape</em><em> (</em><em>n_parameters</em><em>,</em><em>)</em><em>, </em><em>default=</em><em>[</em><em>10</em><em>, </em><em>1</em><em>, </em><em>0.1</em><em>, </em><em>0.01</em><em>, </em><em>0.001</em><em>]</em>) – parameter of interest.</p></li>
<li><p><strong>umap_para</strong> (<em>{array-like}</em><em>, </em><em>default=</em><em>[</em><em>5</em><em>, </em><em>10</em><em>]</em>) – the first number of the array stands for the number of neighbors and
the second number of the array is the total dimension of UMAP</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>self</strong> (<em>object</em>) – Returns self.</p></li>
<li><p><strong>flux_df</strong> (<em>{pandas.DataFrame} of shape (n_reactions, n_samples), optional</em>) – The data table of flux predictions generated from either CFR or DFA.
Only provided if <cite>self.sel_para</cite> is given in the class attributes.</p></li>
<li><p><strong>scores</strong> (<em>{pandas.DataFrame} of shape (n_parameters, 4), optional</em>) – The table of clustering evaluations which comes with coverage rate of HDBSCAN,
rand index, mutual information, and silhouette scores.
It is only provided if <cite>self.sel_para</cite> is not given in the class attributes.
A .csv file with the evaluations is output at the same time.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.fluxRecon">
<span class="sig-name descname"><span class="pre">fluxRecon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">objNorm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clustering</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">validation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.fluxRecon" title="Permalink to this definition">¶</a></dt>
<dd><p>Flux reconstruction</p>
<p>Predict the fluxes by  coefficients*unconstrained models</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>objNorm</strong> (<em>{bool}</em><em>, </em><em>default=True</em>) – normalize the flux values with objective values</p></li>
<li><p><strong>clustering</strong> (<em>{bool}</em><em>, </em><em>default=True</em>) – enable dimension reduction methods or clustering methods for metabolic objectives.
UMAP is the default method in the class, but it is allowed to be swapped by
PCA, tSNE, and PHATE.</p></li>
<li><p><strong>validation</strong> (<em>{bool}</em><em>, </em><em>default=False</em>) – enable the evaluation of regression models with R2, Pearson and Spearman correlations.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>self</strong> (<em>object</em>) – Returns self.</p></li>
<li><p><strong>res</strong> (<em>{pandas.DataFrame}, optional</em>) – a table of evaluation results. Only provided if <cite>validation</cite> is set true.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.functionalAnalysis">
<span class="sig-name descname"><span class="pre">functionalAnalysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cellTypes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sheetnames</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expression_tb_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gesa_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pluri_gene_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sigGenes_root_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.functionalAnalysis" title="Permalink to this definition">¶</a></dt>
<dd><p>Functional analysis overlaying with single-cell data</p>
<p>Refer to functional genes from MSigDB for hypothesis validation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cellTypes</strong> (<em>{string array}</em><em>,</em>) – names of the labels used for analysis. The array should contain unique elements.</p></li>
<li><p><strong>sheetnames</strong> (<em>{string array}</em><em>,</em>) – sheetnames to access the data of significant genes corresponding to <cite>cellTypes</cite>.</p></li>
<li><p><strong>expression_tb_path</strong> (<em>{string}</em><em>,</em>) – path to access the file of gene expression data.</p></li>
<li><p><strong>gesa_name</strong> (<em>{string}</em><em>,</em>) – directory to access all the functional genes.</p></li>
<li><p><strong>pluri_gene_path</strong> (<em>{string}</em><em>,</em>) – path to access the directories where save different types of functional genes.</p></li>
<li><p><strong>sigGenes_root_path</strong> (<em>{string}</em><em>,</em>) – path to access the files of significant genes.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>self</strong> – Returns self.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>object</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This method is only applicable for specific structure of directories and data in Version 0.0.1.</p>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># umap clustering for cluster separation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cf</span> <span class="o">=</span> <span class="n">clustering_func</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>            <span class="n">coef_sel</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="n">coef_sel</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
<span class="gp">&gt;&gt;&gt; </span>            <span class="s1">&#39;./result_figs/&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>            <span class="sa">f</span><span class="s1">&#39;scEmbryo_coef_for_gsea&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>            <span class="n">mets_category</span><span class="p">(),</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># get umap components</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">umaps</span> <span class="o">=</span> <span class="n">cf</span><span class="o">.</span><span class="n">reduction_scatter</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">coef_sel</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">continuous</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">func</span><span class="o">=</span><span class="s1">&#39;UMAP&#39;</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># cell types</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cells</span> <span class="o">=</span> <span class="n">umaps</span><span class="p">[</span><span class="s1">&#39;Cell type&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">exptb</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s1">&#39;/home/daweilin/StemCell/Project_mESC_JinZhang/validation_data/GSE136714/GSE136714_raw.xlsx&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># cluster the umap results</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">recluster</span> <span class="o">=</span> <span class="n">cf</span><span class="o">.</span><span class="n">reclustering</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">umaps</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;UMAP1&#39;</span><span class="p">,</span> <span class="s1">&#39;UMAP2&#39;</span><span class="p">],</span> <span class="n">min_size</span><span class="o">=</span><span class="mi">5</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># functional analysis</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">GESA_analyais</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">umaps</span><span class="p">,</span> <span class="n">recluster</span><span class="p">,</span> <span class="s1">&#39;Cell type&#39;</span><span class="p">,</span> <span class="n">exptb</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">gesa_name</span><span class="o">=</span><span class="s1">&#39;pluripotency&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">cellTypes</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;1C2C&#39;</span><span class="p">,</span> <span class="s1">&#39;2CBC&#39;</span><span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">sheetnames</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;2C&#39;</span><span class="p">,</span> <span class="s1">&#39;BC&#39;</span><span class="p">],</span> <span class="n">reduction_func</span><span class="o">=</span><span class="s1">&#39;UMAP&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">pluri_gene_path</span><span class="o">=</span><span class="s1">&#39;./mcdb/&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">sigGenes_root_path</span><span class="o">=</span><span class="s1">&#39;./data/scEmbryo/GSE136714/sigGenes/&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">save_root_path</span><span class="o">=</span><span class="s1">&#39;./result_figs/&#39;</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.getSpecificSamples">
<span class="sig-name descname"><span class="pre">getSpecificSamples</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sel_label_arr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">update_attr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.getSpecificSamples" title="Permalink to this definition">¶</a></dt>
<dd><p>Slice table with columns associated with correspondingly selected labels</p>
<p>Users select specific columns via given labels to chop the data table.
Since the column names are not the same as the label names,
selected labels will compare with the entire array of labels to index the columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>labels</strong> (<em>{array-like}</em><em> of </em><em>shape</em><em> (</em><em>n_columns</em><em>,</em><em>)</em>) – name array that should has the same length of columns</p></li>
<li><p><strong>sel_label_arr</strong> (<em>{array-like}</em>) – names chosen from labels that should be unique</p></li>
<li><p><strong>update_attr</strong> (<em>{bool}</em><em>, </em><em>default=False</em>) – update class attributes of flux/coefficients dataframe and labels
if update_attr is set true</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>self</strong> – returns self</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>object</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.get_coef">
<span class="sig-name descname"><span class="pre">get_coef</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metType_cluster</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.get_coef" title="Permalink to this definition">¶</a></dt>
<dd><p>Load inferred metabolic objectives.</p>
<p>The coefficients of metabolic objectives were obtained by
the unconstrained models with single objectives regressiing on
the condition/cell-specific constrained models</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>kappa_arr</strong> (<em>{array-like}</em><em> of </em><em>shape</em><em> (</em><em>n_parameters</em><em>,</em><em>)</em><em>, </em><em>default=</em><em>[</em><em>10</em><em>, </em><em>1</em><em>, </em><em>0.1</em><em>, </em><em>0.01</em><em>, </em><em>0.001</em><em>]</em>) – parameter of interest.</p></li>
<li><p><strong>rho_arr</strong> (<em>{array-like}</em><em> of </em><em>shape</em><em> (</em><em>n_parameters</em><em>,</em><em>)</em><em>, </em><em>default=</em><em>[</em><em>10</em><em>, </em><em>1</em><em>, </em><em>0.1</em><em>, </em><em>0.01</em><em>, </em><em>0.001</em><em>]</em>) – parameter of interest.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>self</strong> – Returns self.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>object</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.get_flux">
<span class="sig-name descname"><span class="pre">get_flux</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kappa</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rho</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stack_model</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.get_flux" title="Permalink to this definition">¶</a></dt>
<dd><p>Load inferred metabolic objectives.</p>
<p>The coefficients of metabolic objectives were obtained by
the unconstrained models with single objectives regressiing on
the condition/cell-specific constrained models</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>kappa_arr</strong> (<em>{array-like}</em><em> of </em><em>shape</em><em> (</em><em>n_parameters</em><em>,</em><em>)</em><em>, </em><em>default=</em><em>[</em><em>10</em><em>, </em><em>1</em><em>, </em><em>0.1</em><em>, </em><em>0.01</em><em>, </em><em>0.001</em><em>]</em>) – parameter of interest.</p></li>
<li><p><strong>rho_arr</strong> (<em>{array-like}</em><em> of </em><em>shape</em><em> (</em><em>n_parameters</em><em>,</em><em>)</em><em>, </em><em>default=</em><em>[</em><em>10</em><em>, </em><em>1</em><em>, </em><em>0.1</em><em>, </em><em>0.01</em><em>, </em><em>0.001</em><em>]</em>) – parameter of interest.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>self</strong> – Returns self.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>object</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.get_randomObj_coef">
<span class="sig-name descname"><span class="pre">get_randomObj_coef</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">th</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.get_randomObj_coef" title="Permalink to this definition">¶</a></dt>
<dd><p>get random coefficients of designated objectives</p>
<p>Enable randomly sampling coefficients of selected metabolites based on how many portion
of cells choose a metabolite as a feature. The coefficients will be saved into a .csv file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>th</strong> (<em>{float}</em><em>, </em><em>default=50</em>) – the threshold to select metabolites as features. the number represents the percentage of
cells that choose the metabolites; thus, it ranges from [0, 100]</p></li>
<li><p><strong>sample_size</strong> (<em>{integer}</em><em>, </em><em>default=100000</em>) – number of random samples used to generate coefficients.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.get_unconstrained_model">
<span class="sig-name descname"><span class="pre">get_unconstrained_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sel_mets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flux_plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">corr_plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obj_sanity_check</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.get_unconstrained_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Sanity checks and analysis of unconstrained models</p>
<p>Three goals of the function:
1) Output flux plots of the unconstrained models for publications and sanity check
2) Compute correlations of unconstrained models of metabolites
3) Sanity checks of objective values which should align with the corresponding maximized values</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sel_mets</strong> (<em>{array-like}</em><em>, </em><em>default=</em><em>[</em><em>]</em><em>,</em>) – selected metabolites for publishing tables.
It will output the entire tables if leaving the parameter default.</p></li>
<li><p><strong>flux_plots</strong> (<em>{bool}</em><em>, </em><em>default=False</em><em>,</em>) – Plot the table of fluxes for selected metabolites with clustermaps.</p></li>
<li><p><strong>corr_plots</strong> (<em>{bool}</em><em>, </em><em>default=False</em><em>,</em>) – Plot the correlation of fluxes between unconstrained models with clustermaps.</p></li>
<li><p><strong>obj_sanity_check</strong> (<em>{bool}</em><em>, </em><em>default=False</em><em>,</em>) – Make plots of objective values of each unconstrained model with clustermaps.
It separates the plots based on the category of metabolites.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>self</strong> (<em>object</em>) – returns self</p></li>
<li><p><strong>uncon_res</strong> (<em>{pandas.DataFrame} of shape (n_reactions, n_metabolites),</em>) – The data table of flux predictions generated from either pFBA or FBA without constraints.</p></li>
<li><p><strong>uncon_res_tmp</strong> (<em>{pandas.DataFrame} of shape (n_innercellular_reactions, n_metabolites),</em>) – The data table of flux predictions is actually the <cite>uncon_res</cite> without exchange reactions.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.label_setup">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">label_setup</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_func</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.label_setup" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.output_table">
<span class="sig-name descname"><span class="pre">output_table</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.output_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Export dataframe into a .csv file for publications or other usages</p>
<p>Save .csv files of fluxes or coefficients</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>self</strong> (<em>{object}</em>) – the function will gather the flux data and/or coefficient data for outputs</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>self</strong> – returns self</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>object</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.paraScanFlux">
<span class="sig-name descname"><span class="pre">paraScanFlux</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kappa_arr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[10,</span> <span class="pre">1,</span> <span class="pre">0.1,</span> <span class="pre">0.01,</span> <span class="pre">0.001]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rho_arr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[10,</span> <span class="pre">1,</span> <span class="pre">0.1,</span> <span class="pre">0.01,</span> <span class="pre">0.001]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umap_para</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[5,</span> <span class="pre">10]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.paraScanFlux" title="Permalink to this definition">¶</a></dt>
<dd><p>Scan parameters for the evaluation of clustering/separations of fluxes</p>
<p>This function is only applicable when the flux data <cite>self.flux_df</cite> is loaded.
It only performs a for-loop to scan parameters and save the evaluation results.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>kappa_arr</strong> (<em>{array-like}</em><em> of </em><em>shape</em><em> (</em><em>n_parameters</em><em>,</em><em>)</em><em>, </em><em>default=</em><em>[</em><em>10</em><em>, </em><em>1</em><em>, </em><em>0.1</em><em>, </em><em>0.01</em><em>, </em><em>0.001</em><em>]</em>) – parameter of interest.</p></li>
<li><p><strong>rho_arr</strong> (<em>{array-like}</em><em> of </em><em>shape</em><em> (</em><em>n_parameters</em><em>,</em><em>)</em><em>, </em><em>default=</em><em>[</em><em>10</em><em>, </em><em>1</em><em>, </em><em>0.1</em><em>, </em><em>0.01</em><em>, </em><em>0.001</em><em>]</em>) – parameter of interest.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>self</strong> (<em>object</em>) – Returns self.</p></li>
<li><p><strong>clustering_score_df</strong> (<em>{pandas.DataFrame}</em>)</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The function will update the evaluation score for clustering in class attribute</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.save_clustering_result">
<span class="sig-name descname"><span class="pre">save_clustering_result</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">extra_suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.save_clustering_result" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="SCOOTI.metObjAnalyzer.metObjAnalyzer.tradeoff_analysis">
<span class="sig-name descname"><span class="pre">tradeoff_analysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'coef'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">corr_th</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tri_mets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['gh',</span> <span class="pre">'chsterol',</span> <span class="pre">'glutathione']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pareto_mets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['gh']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pairplot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pairScatter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">triangle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pareto3D</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pareto2DAnalysis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">archetypeAnalysis_w_control</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">archetypeAnalysis_RSS</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">archetypeAnalysis_PCA</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_trait_func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_trait_func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allocation_norm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">row_norm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#SCOOTI.metObjAnalyzer.metObjAnalyzer.tradeoff_analysis" title="Permalink to this definition">¶</a></dt>
<dd><p>Analysis of inferred metabolic objectives.</p>
<p>The coefficients of metabolic objectives were analyzed by
1) correlation of coefficients: a pair of metabolites that compete for allocations,
2) triangle plot: tradeoffs and relationships among three metabolites,
3) pareto analysis: the realistic allocations compared to the optimal solutions in 2D or 3D,
4) trait analysis: PCA-reduced allocations used to identify metabolic traits</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_type</strong> (<em>{string}</em><em>, </em><em>default='coef'</em>) – the data used to perform trade-off analysis. Options include ‘coef’ and ‘fluxRecon’.
The first type of option will get <cite>coef_df</cite> from the class attribute.
The second type of option will get <cite>wobjs</cite> from the class attribute.</p></li>
<li><p><strong>corr_th</strong> (<em>{float}</em><em>, </em><em>default=0.5</em>) – Threshold value used to filter out pairs of metabolites with low magnitude of correlations.</p></li>
<li><p><strong>tri_mets</strong> (<em>{string array}</em><em>, </em><em>default=</em><em>[</em><em>'gh'</em><em>, </em><em>'chsterol'</em><em>, </em><em>'glutathione'</em><em>]</em>) – Input of metabolites for the triangle plot. The length of the input array has to be 3.
<cite>glutathione</cite>, the label, is only applicable when using Recon1 or Shen model which merges
the coefficients of oxidized and reduced glutathione (<cite>gthox</cite> and <cite>gthrd</cite>).
Only useful when the parameter <cite>triangle</cite> is set true.</p></li>
<li><p><strong>pareto_mets</strong> (<em>{string array}</em><em>, </em><em>default=</em><em>[</em><em>'gh'</em><em>]</em>) – Input of metabolites for the plots of i) scatters of relationships or ii) Pareto fronts between pairs of metabolites.
The length of the input array has to be at least 1.</p></li>
<li><p><strong>triangle</strong> (<em>{bool}</em><em>, </em><em>default=False</em>) – enable triangle plots of allocations among the three metabolites defined in <cite>tri_mets</cite>.</p></li>
<li><p><strong>pareto3D</strong> (<em>{bool}</em><em>, </em><em>default=False</em>) – enable Pareto analysis among three metabolites provided in the array of <cite>pareto_mets</cite>
which includes the calculation of Euclidean distance from the data points to the surface.</p></li>
<li><p><strong>traitAnalysis</strong> (<em>{bool}</em><em>, </em><em>default=False</em>) – enable PCA-based analysis of allocations in high dimensional objective coefficients.
The vertices of traits represent the extreme traits of the metabolic objectives
based on the mechanistic limitation.</p></li>
<li><p><strong>sample_trait_func</strong> (<em>{bool}</em><em>, </em><em>default=False</em>) – additional function that gets omics-based data points closer to traits on vertices</p></li>
<li><p><strong>control_trait_func</strong> (<em>{bool}</em><em>, </em><em>default=False</em>) – additional function that gets mechanistic data points closet to traits on vertices</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>self</strong> – Returns self.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>object</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">SCOOTI</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="regressionAnalyzer.html">Welcome to the documentation of metabolic objective analysis</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Welcome to the documentation of metabolic objective analysis</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#module-SCOOTI.metObjAnalyzer">Core Analysis Module</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quick Start</a></li>
<li class="toctree-l1"><a class="reference internal" href="scootisettings.html">SCOOTI Settings</a></li>
<li class="toctree-l1"><a class="reference internal" href="advancedusage.html">Advanced usage</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="regressionAnalyzer.html" title="previous chapter">Welcome to the documentation of metabolic objective analysis</a></li>
      <li>Next: <a href="quickstart.html" title="next chapter">Quick Start</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2023, Chandrasekaran Lab, University of Michigan.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.3.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.13</a>
      
      |
      <a href="_sources/metObjAnalyzer.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>