
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Quick Start &#8212; SCOOTI 0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="SCOOTI Settings" href="scootisettings.html" />
    <link rel="prev" title="Welcome to SCOOTI’s documentation!" href="index.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="quick-start">
<h1>Quick Start<a class="headerlink" href="#quick-start" title="Permalink to this heading">¶</a></h1>
<section id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this heading">¶</a></h2>
<p><strong>SCOOTI: Single-Cell Optimization Objective and Tradeoff Inference</strong> is a data-driven computational framework to identify cell-specific objectives and trade-offs from transcriptomics data.</p>
<p><strong>SCOOTI</strong> is able to</p>
<ol class="arabic simple">
<li><p>Formulate metabolic objectives for various cell types, such as cancer cell lines and embryos, in different conditions.</p></li>
<li><p>Incorporate only transcriptomics data to identify metabolic objectives that is sufficient, but other omics data also works.</p></li>
<li><p>Extract information of metabolism with cell-specific objectives to analyze the difference/similarity of each cell fate.</p></li>
<li><p>Analyze the relationships among metabolites and integrate mechanistic trade-offs to identify metabolic traits.</p></li>
</ol>
<p><strong>What is objective function in metabolic modeling?</strong></p>
<p>Biologically, metabolic objective is the metabolic goal a cell wants to achieve. Take bacterial cells for example, scientists working on metabolic modeling usually assumed that growth/doubling is the most important goal for the cells. They built a objective function called <cite>biomass</cite> consisted of nucleotides, amino acides, lipids and so on to represent the growth/doubling of cells which sometimes considered as the phenotypes of the cells.</p>
<p>Mathematically, objective function is a linear combination of several different metabolites. For example,
<code class="docutils literal notranslate"><span class="pre">Objective</span> <span class="pre">=</span> <span class="pre">3</span> <span class="pre">ATP</span> <span class="pre">+</span> <span class="pre">2</span> <span class="pre">NADPH</span> <span class="pre">+</span> <span class="pre">GSH</span> <span class="pre">+</span> <span class="pre">Ala</span></code>
where <strong>ATP</strong>, <strong>NADPH</strong>, <strong>GSH</strong>, and <strong>Ala</strong> are metabolites contributing to the metabolic objective. In other words, the phenotypes of cells or the “activities” of cells are determined by the magnitude of cells.</p>
<p><strong>We assume</strong></p>
<ol class="arabic simple">
<li><p>Every cell can optimize their functions to achieve certain metabolic goals depending on different conditions.</p></li>
<li><p>Biomass, namely cell growth/doubling, is not always the goal for different cells, especially for mammalian cells.</p></li>
<li><p>Different metabolic distributions and expression profiles can lead to the same metabolic objectives.</p></li>
</ol>
</section>
<section id="system-requirements">
<h2>System Requirements<a class="headerlink" href="#system-requirements" title="Permalink to this heading">¶</a></h2>
<section id="os-requirements">
<h3>OS Requirements<a class="headerlink" href="#os-requirements" title="Permalink to this heading">¶</a></h3>
<p>The framework has been tested on the following systems:</p>
<ul class="simple">
<li><p>Linux: Centos Linux 7 (core) and Ubuntu 22.04</p></li>
</ul>
</section>
<section id="python-dependencies">
<h3>Python dependencies<a class="headerlink" href="#python-dependencies" title="Permalink to this heading">¶</a></h3>
<ul class="simple">
<li><p>numpy</p></li>
<li><p>scipy</p></li>
<li><p>pandas</p></li>
<li><p>matplotlib</p></li>
<li><p>seaborn</p></li>
</ul>
</section>
<section id="matlab-dependencies">
<h3>MATLAB dependencies<a class="headerlink" href="#matlab-dependencies" title="Permalink to this heading">¶</a></h3>
<ul class="simple">
<li><p>cplex (optional)</p></li>
<li><p>gurobi</p></li>
<li><p>cobratoolbox</p></li>
</ul>
</section>
<section id="installation">
<h3>Installation<a class="headerlink" href="#installation" title="Permalink to this heading">¶</a></h3>
<p>To use SCOOTI, first install it using pip:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp gp-VirtualEnv">(venv)</span> <span class="gp">$ </span>pip<span class="w"> </span>install<span class="w"> </span>git+https://github.com/dwgoblue/SCOOTI.git<span class="w"> </span>--upgrade
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Cobratoolbox is required for SCOOTI to model contrained and unconstrained models. The instruction of cobratoolbox and its installation can be found <a class="reference external" href="https://opencobra.github.io/cobratoolbox/stable/installation.html">https://opencobra.github.io/cobratoolbox/stable/installation.html</a>. In addition, optimization solvers are required for cobratoolbox. We recommend Gurobi solver for the linear version of iMAT and MOOMIN, and CPLEX 20.0.1 for COMPASS.</p>
</div>
</section>
</section>
<section id="unconstrained-models-with-single-objectives">
<h2>Unconstrained Models with single objectives<a class="headerlink" href="#unconstrained-models-with-single-objectives" title="Permalink to this heading">¶</a></h2>
<p>Unconstrained models optimized the demand reactions of single metabolites across different compartments. For example, <code class="docutils literal notranslate"><span class="pre">atp[c]+atp[n]+atp[m]--&gt;</span></code> is one of the single objective functions. Optimizing this function is equivalent to maximizing the production of ATP. The bash script below shows the example to generate 52 unconstrained models which optimize 52 different single metabolites recorded in the csv file.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="ch">#!/bin/bash</span>
<span class="c1"># input settings</span>
<span class="c1"># path to access your matlab-version cobratoolbox</span>
<span class="nv">COBRA_path</span><span class="o">=</span><span class="s1">&#39;./cobratoolbox/&#39;</span>
<span class="c1"># path to access the metabolic model</span>
<span class="nv">GEM_path</span><span class="o">=</span><span class="s1">&#39;./GEMs/Shen2019.mat&#39;</span>
<span class="c1"># name of the model</span>
<span class="nv">model_name</span><span class="o">=</span><span class="s1">&#39;Recon1&#39;</span>
<span class="c1"># leave it blank if no user-defined objectives</span>
<span class="nv">obj_candidate_list_file</span><span class="o">=</span><span class="s1">&#39;./obj52_metabolites_recon1.csv&#39;</span>

<span class="c1"># path to access the significant genes data</span>
<span class="nv">data_dir</span><span class="o">=</span><span class="s1">&#39;./sigGenes/prolif_qui/&#39;</span>
<span class="nv">prefix_name</span><span class="o">=</span><span class="s1">&#39;model&#39;</span><span class="w"> </span><span class="c1"># name of the experiment pls set to &#39;model&#39; for unconstraint models</span>
<span class="nv">medium</span><span class="o">=</span><span class="s1">&#39;DMEMF12&#39;</span><span class="w"> </span><span class="c1"># KSOM for embryos and DMEMF12 for cell culture</span>
<span class="nv">save_root_path</span><span class="o">=</span><span class="s1">&#39;./fluxPrediction/unconstrained_models/pfba/&#39;</span><span class="w"> </span><span class="c1"># path to save predicted fluxes</span>

<span class="c1"># start the simulation of flux predictions</span>
matlab<span class="w"> </span>-nosplash<span class="w"> </span>-noFigureWindows<span class="w"> </span>-r<span class="w"> </span><span class="s2">&quot;multiObj_CBM(~, </span><span class="nv">$COBRA_path</span><span class="s2">, </span><span class="nv">$GEM_path</span><span class="s2">, </span><span class="nv">$model_name</span><span class="s2">, </span><span class="nv">$obj_candidate_list_file</span><span class="s2">, </span><span class="nv">$data_dir</span><span class="s2">, </span><span class="nv">$prefix_name</span><span class="s2">, </span><span class="nv">$medium</span><span class="s2">, </span><span class="nv">$save_root_path</span><span class="s2">)&quot;</span>
</pre></div>
</div>
</section>
<section id="omics-based-constrained-models">
<h2>Omics-based Constrained Models<a class="headerlink" href="#omics-based-constrained-models" title="Permalink to this heading">¶</a></h2>
<p>Constrained models are predicted metabolic fluxes with respect to transcriptomics data. This reflects the control of the metabolic genes. Here, we show the example to generate constrained models for proliferative and quiescent states in a bash file.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="ch">#!/bin/bash</span>
<span class="c1"># input settings</span>
<span class="c1"># path to access your matlab-version cobratoolbox</span>
<span class="nv">COBRA_path</span><span class="o">=</span><span class="s1">&#39;./cobratoolbox/&#39;</span>
<span class="c1"># path to access the metabolic model</span>
<span class="nv">GEM_path</span><span class="o">=</span><span class="s1">&#39;./GEMs/Shen2019.mat&#39;</span>
<span class="c1"># name of the model</span>
<span class="nv">model_name</span><span class="o">=</span><span class="s1">&#39;Recon1&#39;</span>
<span class="c1"># leave it blank if no user-defined objectives</span>
<span class="nv">obj_candidate_list_file</span><span class="o">=</span><span class="s1">&#39;./obj52_metabolites_recon1.csv&#39;</span>
<span class="c1"># objective values</span>
<span class="nv">input_obj_tb</span><span class="o">=</span><span class="s1">&#39;&#39;</span>

<span class="c1"># parameter settings</span>
<span class="nv">DFA_kappa</span><span class="o">=</span>-1
<span class="nv">CFR_kappa</span><span class="o">=</span><span class="m">0</span>.1
<span class="nv">CFR_rho</span><span class="o">=</span><span class="m">10</span>
<span class="nv">paraLen</span><span class="o">=</span><span class="m">1</span><span class="w"> </span><span class="c1"># how many kappa/rho used for scanning</span>
<span class="nv">random_para</span><span class="o">=</span><span class="m">0</span><span class="w"> </span><span class="c1"># bool, 1 to enable random sampling</span>
<span class="nv">init_objective</span><span class="o">=</span><span class="m">1</span><span class="w"> </span><span class="c1"># 1 for none, 2 for biomass objective</span>
<span class="nv">genekoflag</span><span class="o">=</span><span class="m">0</span><span class="w"> </span><span class="c1"># bool</span>
<span class="nv">rxnkoflag</span><span class="o">=</span><span class="m">0</span><span class="w"> </span><span class="c1"># bool</span>
<span class="nv">FVAflag</span><span class="o">=</span><span class="m">0</span><span class="w"> </span><span class="c1"># bool</span>
<span class="nv">pfba</span><span class="o">=</span><span class="m">1</span><span class="w"> </span><span class="c1"># 0 for fba and 1 for pfba (minimize sum of fluxes)</span>
<span class="nv">medium_perturbation</span><span class="o">=</span><span class="m">0</span><span class="w"> </span><span class="c1"># 1 for depletion or excess of metabolites in medium</span>
<span class="nv">pairwise_CFR_model</span><span class="o">=</span><span class="m">0</span>
<span class="nv">algorithm</span><span class="o">=</span><span class="s1">&#39;iMAT&#39;</span>

<span class="c1"># path to access the significant genes data</span>
<span class="nv">data_dir</span><span class="o">=</span><span class="s1">&#39;./sigGenes/prolif_qui/&#39;</span>
<span class="nv">prefix_name</span><span class="o">=</span><span class="s1">&#39;QuiProlif&#39;</span><span class="w"> </span><span class="c1"># name of the experiment pls set to &#39;model&#39; for unconstraint models</span>
<span class="nv">medium</span><span class="o">=</span><span class="s1">&#39;DMEMF12&#39;</span><span class="w"> </span><span class="c1"># KSOM for embryos and DMEMF12 for cell culture</span>
<span class="nv">late_stage</span><span class="o">=</span><span class="s1">&#39;upgenes&#39;</span><span class="w"> </span><span class="c1"># suffix of the file names of significant up-genes</span>
<span class="nv">early_stage</span><span class="o">=</span><span class="s1">&#39;dwgenes&#39;</span><span class="w"> </span><span class="c1"># suffix of the file names of significant down-genes</span>
<span class="nv">simulation</span><span class="o">=</span><span class="s1">&#39;CFR&#39;</span><span class="w"> </span><span class="c1"># CFR for transcriptomics and proteomics; DFA for metabolomics</span>
<span class="nv">constraint</span><span class="o">=</span><span class="m">1</span><span class="w"> </span><span class="c1"># apply constraints to the model</span>
<span class="nv">save_root_path</span><span class="o">=</span><span class="s1">&#39;./fluxPrediction/prolif_qui/&#39;</span><span class="w"> </span><span class="c1"># path to save predicted fluxes</span>
<span class="nv">CFR_model_path</span><span class="o">=</span><span class="s1">&#39;&#39;</span>
<span class="nv">extraWeight</span><span class="o">=</span><span class="m">0</span>

<span class="c1"># for-loop settings</span>
<span class="c1"># if scanning kappa from 1E-3 to 1E1 and rho from 1E-3 to 1E1,</span>
<span class="c1"># there will be 25 combinations in total.</span>
<span class="nv">START_NUM</span><span class="o">=</span><span class="m">1</span>
<span class="nv">END_NUM</span><span class="o">=</span><span class="nv">$paraLen</span>*<span class="nv">$paraLen</span>

<span class="c1"># execute parameter scanning</span>
<span class="k">for</span><span class="w"> </span><span class="o">((</span><span class="w"> </span><span class="nv">run</span><span class="o">=</span><span class="nv">$START_NUM</span><span class="p">;</span><span class="w"> </span>run&lt;<span class="o">=</span>END_NUM<span class="p">;</span><span class="w"> </span>run++<span class="w"> </span><span class="o">))</span><span class="p">;</span><span class="w"> </span><span class="k">do</span>
<span class="w">  </span>matlab<span class="w"> </span>-nosplash<span class="w"> </span>-noFigureWindows<span class="w"> </span>-r<span class="w"> </span><span class="s2">&quot;multiObj_CBM(</span><span class="nv">$run</span><span class="s2">, </span><span class="nv">$DFA_kappa</span><span class="s2">, </span><span class="nv">$CFR_kappa</span><span class="s2">, </span><span class="nv">$CFR_rho</span><span class="s2">, </span><span class="nv">$COBRA_path</span><span class="s2">,</span><span class="nv">$GEM_path</span><span class="s2">, </span><span class="nv">$model_name</span><span class="s2">, </span><span class="nv">$obj_candidate_list_file</span><span class="s2">, </span><span class="nv">$input_obj_tb</span><span class="s2">, </span><span class="nv">$paraLen</span><span class="s2">, </span><span class="nv">$random_para</span><span class="s2">, </span><span class="nv">$init_objective</span><span class="s2">, </span><span class="nv">$genekoflag</span><span class="s2">, </span><span class="nv">$rxnkoflag</span><span class="s2">, </span><span class="nv">$FVAflag</span><span class="s2">, </span><span class="nv">$pfba</span><span class="s2">, </span><span class="nv">$medium_perturbation</span><span class="s2">, </span><span class="nv">$data_dir</span><span class="s2">, </span><span class="nv">$prefix_name</span><span class="s2">, </span><span class="nv">$medium</span><span class="s2">, </span><span class="nv">$late_stage</span><span class="s2">, </span><span class="nv">$early_stage</span><span class="s2">, </span><span class="nv">$simulation</span><span class="s2">, </span><span class="nv">$constraint</span><span class="s2">, </span><span class="nv">$save_root_path</span><span class="s2">, </span><span class="nv">$CFR_model_path</span><span class="s2">, </span><span class="nv">$pairwise_CFR_model</span><span class="s2">, </span><span class="nv">$extraWeight</span><span class="s2">, </span><span class="nv">$algorithm</span><span class="s2">)&quot;</span>
<span class="k">done</span>
</pre></div>
</div>
</section>
<section id="inference-of-metabolic-objectives">
<h2>Inference of Metabolic Objectives<a class="headerlink" href="#inference-of-metabolic-objectives" title="Permalink to this heading">¶</a></h2>
<p>After we prepared the constrained models and unconstrained models, we are able to infer metabolic objectives. Here, we show the example to infer metabolic objectives with meta-learner models for proliferative and quiescent states in a bash file.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># General settings</span>
<span class="nv">unconModel</span><span class="o">=</span>./fluxPrediction/unconstrained_models/pfba/DMEMF12/
<span class="nv">conModel</span><span class="o">=</span>./fluxPrediction/prolif_qui/
<span class="nv">savePath</span><span class="o">=</span>./regression_models/QuiProlif/
<span class="nv">geneListPath</span><span class="o">=</span>./unique_gene_list.mat
<span class="nv">kappa</span><span class="o">=</span><span class="m">0</span>.1
<span class="nv">rho</span><span class="o">=</span><span class="m">10</span>
<span class="c1"># run model</span>
python<span class="w"> </span>SCOOTI_trainer.py<span class="w"> </span>--unconModel<span class="w"> </span><span class="nv">$unconModel</span><span class="w"> </span>--conModel<span class="w"> </span><span class="nv">$conModel</span><span class="w"> </span>--savePath<span class="w"> </span><span class="nv">$savePath</span><span class="w"> </span>--kappaArr<span class="w"> </span><span class="nv">$kappa</span><span class="w"> </span>--rhoArr<span class="w"> </span><span class="nv">$rho</span><span class="w"> </span>--expName<span class="w"> </span>QuiProlif<span class="w"> </span>--unconNorm<span class="w"> </span>T<span class="w"> </span>--conNorm<span class="w"> </span>F<span class="w"> </span>--medium<span class="w"> </span>DMEMF12<span class="w"> </span>--method<span class="w"> </span>CFR<span class="w"> </span>--model<span class="w"> </span>recon1<span class="w"> </span>--inputType<span class="w"> </span>flux<span class="w"> </span>--rank<span class="w"> </span>F<span class="w"> </span>--stackModel<span class="w"> </span>F<span class="w"> </span>--sampling<span class="w"> </span>F<span class="w"> </span>--learner<span class="w"> </span>L<span class="w"> </span>--geneKO<span class="w"> </span>F<span class="w"> </span>--geneListPath<span class="w"> </span><span class="nv">$geneListPath</span>
</pre></div>
</div>
</section>
<section id="metabolic-objective-analysis">
<h2>Metabolic Objective Analysis<a class="headerlink" href="#metabolic-objective-analysis" title="Permalink to this heading">¶</a></h2>
<p>Once we got the metabolic objectives, we can rely on a Python class <cite>metObjAnalyzer</cite> to perform statistical analysis, phenotype analysis, trade-off analysis, and so on. Here, we show the example to analyze the inferred metabolic objectives for proliferative and quiescent states in a python script.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="ch">#!/usr/bin/env python</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="c1"># initiate a instance with a python object</span>
<span class="n">moa</span> <span class="o">=</span> <span class="n">metObjAnalyzer</span><span class="p">(</span>
            <span class="n">flux_paths</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;QP&#39;</span><span class="p">:</span><span class="s1">&#39;./fluxPrediction/prolif_qui/&#39;</span><span class="p">,</span>
                    <span class="p">},</span>
            <span class="n">coef_paths</span><span class="o">=</span><span class="p">{</span>
    <span class="s1">&#39;QP&#39;</span><span class="p">:</span><span class="sa">f</span><span class="s1">&#39;./regression_models/QuiProlif_paraScan/flux_sl_BulkRNAseq_QuiProlif_input_norm_outcome_nonorm_k0.1_r10.csv&#39;</span><span class="p">,</span>
                <span class="p">},</span>
            <span class="n">save_root_path</span><span class="o">=</span><span class="s1">&#39;./result_files/&#39;</span><span class="p">,</span>
            <span class="n">GEM_path</span><span class="o">=</span><span class="s1">&#39;./metabolicModel/GEMs/Shen2019.mat&#39;</span><span class="p">,</span>
            <span class="n">uncon_model_path</span><span class="o">=</span><span class="s1">&#39;./fluxPrediction/unconstrained_models/pfba/DMEMF12/&#39;</span><span class="p">,</span>
            <span class="n">col_map</span><span class="o">=</span><span class="p">{},</span>
            <span class="n">label_func</span><span class="o">=</span><span class="n">label_func</span><span class="p">,</span>
            <span class="n">samplingFlux_path</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span>
            <span class="n">sel_para</span><span class="o">=</span><span class="s1">&#39;k0.1_r10&#39;</span><span class="p">,</span>
            <span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;QuiProlif&#39;</span><span class="p">,</span>
            <span class="n">medium</span><span class="o">=</span><span class="s1">&#39;DMEMF12&#39;</span><span class="p">,</span>
            <span class="p">)</span>
<span class="c1"># get coefficients of metabolic objectives</span>
<span class="n">moa</span><span class="o">.</span><span class="n">get_coef</span><span class="p">(</span><span class="n">metType_cluster</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="c1"># analyze the coefficients with distance to biomass and statistical comparisons between groups</span>
<span class="n">moa</span><span class="o">.</span><span class="n">coefAnalysis</span><span class="p">(</span>
            <span class="n">norm</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">unknown_clustering</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">clustering</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">entropy</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">distance</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">compare</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">umap_para</span><span class="o">=</span><span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">50</span><span class="p">],</span>
            <span class="n">recluster_min_size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
            <span class="n">method</span><span class="o">=</span><span class="s1">&#39;average&#39;</span><span class="p">,</span>
            <span class="n">ref_col</span><span class="o">=</span><span class="s1">&#39;Proliferative&#39;</span><span class="p">,</span>
            <span class="n">cutoff</span><span class="o">=</span><span class="mf">0.01</span>
            <span class="p">)</span>
</pre></div>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">SCOOTI</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Quick Start</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="#system-requirements">System Requirements</a></li>
<li class="toctree-l2"><a class="reference internal" href="#unconstrained-models-with-single-objectives">Unconstrained Models with single objectives</a></li>
<li class="toctree-l2"><a class="reference internal" href="#omics-based-constrained-models">Omics-based Constrained Models</a></li>
<li class="toctree-l2"><a class="reference internal" href="#inference-of-metabolic-objectives">Inference of Metabolic Objectives</a></li>
<li class="toctree-l2"><a class="reference internal" href="#metabolic-objective-analysis">Metabolic Objective Analysis</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="scootisettings.html">SCOOTI Settings</a></li>
<li class="toctree-l1"><a class="reference internal" href="advancedusage.html">Advanced usage</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="index.html" title="previous chapter">Welcome to SCOOTI’s documentation!</a></li>
      <li>Next: <a href="scootisettings.html" title="next chapter">SCOOTI Settings</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2023, Chandrasekaran Lab, University of Michigan.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.3.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.13</a>
      
      |
      <a href="_sources/quickstart.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>